---
title: 'Using Monocle as input to **tradeSeq**'
author: "Koen Van den Berge and Hector Roux de BÃ©zieux"
bibliography: tradeSeq.bib
date: "9/10/2018"
output: 
  rmarkdown::html_document:
    toc: true
    toc_depth: 3
vignette: >
  %\VignetteIndexEntry{Monocle + tradeSeq}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---
```{r, echo = FALSE}
library(knitr)
# For a smaller vignette
# This can be deleted when trying to follow along
opts_chunk$set(dpi = 58)
```

# Introduction

`tradeSeq` is an `R` package that allows analysis of gene expression along trajectories. While it has been developed and applied to single-cell RNA-sequencing (scRNA-seq) data, its applicability extends beyond that, and also allows the analysis of, e.g., single-cell ATAC-seq data along trajectories or bulk RNA-seq time series datasets. For every gene in the dataset, `tradeSeq` fits a generalized additive model (GAM) by building on the `mgcv` R package. It then allows statistical inference on the GAM by assessing contrasts of the parameters of the fitted GAM model, aiding in interpreting complex datasets. All details about the `tradeSeq` model and statistical tests are described in our preprint [@VandenBerge2019a].

In this vignette, we analyze a subset of the data from [@Paul2015]. 
A `SingleCellExperiment` object of the data has been provided with the [`tradeSeq`](https://github.com/statOmics/tradeSeqpaper) package and can be retrieved as shown below. The data and UMAP reduced dimensions were derived from following the [Monocle 3 vignette](http://cole-trapnell-lab.github.io/monocle-release/monocle3/#tutorial-1-learning-trajectories-with-monocle-3). 
# Load data

```{r, warning=F, message=F}
library(tradeSeq)
library(RColorBrewer)
library(SingleCellExperiment)
library(monocle)
library(dplyr)
library(tidyr)

# For reproducibility
RNGversion("3.5.0")
palette(brewer.pal(8, "Dark2"))
data(countMatrix, package = "tradeSeq")
counts <- as.matrix(countMatrix)
rm(countMatrix)
data(crv, package = "tradeSeq")
data(celltype, package = "tradeSeq")
```

# Fit trajectories using Monocle

We will fit developmental trajectories using the `monocle` package [@Street2018a]. 

```{r, warning=FALSE, out.width="50%", fig.asp=.6}
set.seed(200)
counts <- counts[, 1:1000]
df <- data.frame(cells = colnames(counts), cellType = celltype[1:1000])
rownames(df) <- df$cells
cds <- newCellDataSet(counts, phenoData = new("AnnotatedDataFrame", data = df))
cds <- estimateSizeFactors(cds)
cds <- reduceDimension(cds, max_components = 2, reduction_method = "ICA")
cds <- orderCells(cds)
cds <- orderCells(cds, root_state = 5)
```


We find 3 lineages in this case, instead of the two that are supposed to be in the dataset. The trajectory can be visualized using the `plotGeneCount` function, using either the cluster labels or cell type to color the cells.

```{r}
y_to_cells <- cds@auxOrderingData[[cds@dim_reduce_type]]
y_to_cells <- y_to_cells$pr_graph_cell_proj_closest_vertex %>%
  as.data.frame() %>%
  mutate(cells = rownames(.)) %>%
  rename("Y" = V1)
root <- cds@auxOrderingData[[cds@dim_reduce_type]]$root_cell
root <- y_to_cells$Y[y_to_cells$cells == root]
endpoints <- names(which(igraph::degree(cds@minSpanningTree) == 1))
endpoints <- endpoints[endpoints != paste0("Y_", root)]
mst <- cds@minSpanningTree
cellWeights <- lapply(endpoints, function(endpoint) {
  path <- igraph::shortest_paths(mst, root, endpoint)$vpath[[1]]
  path <- as.character(path)
  df <- y_to_cells %>%
    filter(Y %in% path)
  df <- data.frame(weights = as.numeric(colnames(cds) %in% df$cells))
  colnames(df) <- endpoint
  return(df)
}) %>% bind_cols() 
pseudotime <- purrr::map_df(cellWeights, function(w) cds$Pseudotime)
rownames(cellWeights) <- rownames(pseudotime) <- colnames(cds)
```

# Session

```{r}
sessionInfo()
```

# References
