---
title: 'Vignette for **tradeSeq**: **tra**jectory **d**ifferential **e**xpression analysis for **Seq**uncing data'
author: "Koen Van den Berge and Hector Roux de BÃ©zieux"
bibliography: tradeSeq.bib
date: "9/10/2018"
output: 
  rmarkdown::html_document:
    toc: true
    toc_depth: 2
vignette: >
  %\VignetteIndexEntry{'Vignette for **tradeSeq**}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

In this vignette, we analyse a subset of the data from [@Paul2015]. 
A `SingleCellExperiment` object of the data has been provided with the [`tradeSeq`](https://github.com/statOmics/tradeSeqpaper) package and can be retrieved as shown below. The data and UMAP reduced dimensions were derived from following the [Monocle 3 vignette](http://cole-trapnell-lab.github.io/monocle-release/monocle3/#tutorial-1-learning-trajectories-with-monocle-3). 

# Load data

```{r, warning=F, message=F}
suppressPackageStartupMessages({
  library(SingleCellExperiment)
  library(mgcv)
  library(tradeSeq)
  library(slingshot)
  library(RColorBrewer)
  library(dplyr)
  library(ggplot2)
  library(tidyr)
  library(cowplot)
})
palette(brewer.pal(8,"Dark2"))
download.file(
  "https://github.com/statOmics/tradeSeqPaper/raw/master/data/se_paul.rda",
  destfile = "./se_paul.rda")
load("./se_paul.rda")
```

# Fit trajectories using slingshot

We will fit developmental trajectories using the `slingshot` package [@Street2018a]. `slingshot` requires cluster labels as input, and fits trajectories in reduced dimension. We will use the reduced space calculated with the UMAP method, which is pre-calculated in the `se` object.
We cluster the data using k-means with $7$ clusters. Since we know which cells are the progenitor cell type, we define the starting point for the trajectories as input for `slingshot`. Note that this argument is optional, and not required to run `slingshot`.

```{r, warning=FALSE}
set.seed(200)
rd <- reducedDims(se)$UMAP
cl <- kmeans(rd, centers = 7)$cluster
plot(rd, col = brewer.pal(9, "Set1")[cl], pch = 16, asp = 1,
     cex = 2/3)
library(slingshot)
lin <- getLineages(rd, clusterLabels = cl, start.clus = 6)
crv <- getCurves(lin)
```


We find two lineages for this dataset. The trajectory can be visualized using the `plotGeneCount` function, using either the cluster labels or cell type to color the cells.

```{r}
plotGeneCount(rd = rd, curve = crv, counts = counts, clusters = cl)

celltype <- factor(colData(se)$cell_type2)

plotGeneCount(rd = rd, curve = crv, counts = counts, clusters = celltype, 
              title = "Colored by cell type")
legend("topright", levels(celltype),
       col = brewer.pal(9, "Set1")[1:nlevels(celltype)],
       pch = 16, cex = 1 / 2, bty = "n")
```

# Fit additive models

After estimating the trajectory, we can fit generalized additive models (GAMs) with the `tradeSeq` package. Internally, this package builds on the `mgcv` package by fitting additive models using the `gam` function. The core function from `tradeSeq`, `fitGAM`, will use cubic splines as basis functions, and it tries to ensure that every lineage will end at a knot point of a smoother. By default, we allow for $10$ knots for every lineage, but this can be changed with the `nknots` function. More knots will allow more flexibility, but also increase the risk of overfitting.
By default, the GAM model estimates one smoother for every lineage using the negative binomial distribution. If you want to allow for other fixed effects (e.g. batch effects), then an additional model matrix can be provided with the `U` argument. 

We use the effective library size, estimated with TMM [@Robinson2010], as offset in the model. We allow for alternatives by allowing a user-defined offset with the `offset` argument.

This dataset consists of UMI counts, and we do not expect zero inflation to be a big problem. However, we also allow to fit zero inflated negative binomial (ZINB) GAMs by providing observation-level weights to `fitGAM` using the `weights` argument. The `weights` must correspond to the posterior probability that a count belongs to the count component of the ZINB distribution [@VandenBerge2018].

For the vignette, we fit smoothers for a filtered set of genes in the dataset, 239 genes in total. We also include the *Irf8* gene, since it is a known transcription factor involved in hematopoiesis.

The fitGam function relies on __BiocParallel__ to implement parallelization, progress bars and so on.

```{r}
# Register BiocParallel Serial Execution
BiocParallel::register(BiocParallel::SerialParam())

counts <- assays(se)$counts %>% as.matrix()
filt <- rowSums(counts > 8) > ncol(counts)/100
filt["Irf8"] <- TRUE
counts <- counts[filt, ]
gamList <- fitGAM(counts = counts,
                  pseudotime = slingPseudotime(crv, na = FALSE),
                  cellWeights = slingCurveWeights(crv))
# This takes about 1mn to run
```

One may explore the results of a model by requesting its summary.

```{r}
summary(gamList[["Irf8"]])
```

You can also plot the cells in reduced dimension to see where the knots are located.

```{r}
plotGeneCount(rd = rd, curve = crv, counts = counts, clusters = cl,
              models = gamList)
```

# Within-lineage comparisons

## Association of gene expression with pseudotime

A first exploration of the data analysis may consist in checking whether gene expression is associated with a particular lineage.
The statistical test performed here, implemented in the `associationTest` function, is testing the null hypothesis that all smoother coefficients are equal to each other. This can be interpreted as testing whether the smoothed gene expression is significantly changing along pseudotime.

```{r}
assoRes <- associationTest(gamList)
```

Related, one can extract the p-values generated by the `mgcv` package using the `getSmootherPvalues` function. These p-values are derived from a test that assesses the null hypothesis that all smoother coefficients are equal to zero. Note, however, that their interpretation is thus more complex. A significant lineage for a particular gene might thus be the result of (a) a different mean expression in that lineage as compared to the overall expression of that gene, or (b) significantly varying expression along that lineage, even if the means are equal, or (c) a combination of both.
This function extracts the p-values calculated by `mgcv` from the GAM, and will return `NA` for genes that we were unable to fit properly.
Similarly, the test statistics may be extracted with `getSmootherTestStats`.
Since this dataset was pre-filtered to only contain relevant genes, all p-values (test statistics) will be very low (high).

```{r}
pvalLineage <- getSmootherPvalues(gamList)
statLineage <- getSmootherTestStats(gamList)
```

## Discovering progenitor marker genes

In order to discover marker genes of the progenitor cell population, researchers may be interested in assessing differential expression between the progenitor cell population (i.e., the starting point of a lineage) with the differentiated cell type population (i.e., the end point of a lineage).
In the function `startVsEndTest`, we have implemented a Wald test that tests the null hypothesis that the expression at the starting point of the smoother (progenitor population) is identical to the expression at the end point of the smoother (differentiated population).
The test basically involves a comparison between two smoother coefficients for every lineage.
The function `startVsEndTest` performs an omnibus test across all lineages by default, but you can also assess all lienages separately by setting `lineages=TRUE`.
Below, we adopt an omnibus test across the two lineages.

```{r}
startRes <- startVsEndTest(gamList)
```

We can visualize the estimated smoothers for the most significant gene.

```{r}
oStart <- order(startRes$waldStat, decreasing = TRUE)
sigGeneStart <- names(gamList)[oStart[1]]
plotSmoothers(gamList[[sigGeneStart]])
```

Alternatively, we can color the cells in UMAP space with that gene's expression.

```{r}
plotGeneCount(rd, crv, counts, gene = sigGeneStart)
```


# Between-lineage comparisons

## Discovering differentiated cell type markers

`tradeSeq` can discover marker genes for the differentiated cell types by comparing the end points of the lineage-specific smoothers. This is implemented in the `diffEndTest` function.
By default, `diffEndTest` performs an omnibus test, testing the null hypothesis that the endpoint expression is equal for all lineages using a multivariate Wald test.
If more than two trajectories are present, one can assess all pairwise comparisons using the `pairwise=TRUE` argument.

```{r}
endRes <- diffEndTest(gamList)
```

We can plot the most significant gene using the `plotSmoothers` function.

```{r}
o <- order(endRes$waldStat, decreasing = TRUE)
sigGene <- names(gamList)[o[1]]
plotSmoothers(gamList[[sigGene]])
```

Alternatively, we can color the cells in UMAP space with that gene's expression.

```{r}
plotGeneCount(rd, crv, counts, gene = sigGene)
```

## Discovering genes with different expression patterns

Asides from testing at the level of the differentiated cell type, researchers may be interested in assessing the expression pattern of a gene over pseudotime.
The function `patternTest` implements a statistical method that checks whether the smoothed gene expression is equal along pseudotime between two or multiple lineages. In practice, we use $100$ points, equally distributed along pseudotime, that are compared between two (or multiple) lineages, and this number can be changed using the `nPoints` argument.

```{r}
patternRes <- patternTest(gamList)
oPat <- order(patternRes$waldStat, decreasing = TRUE)
head(rownames(patternRes)[oPat])
plotSmoothers(gamList[[rownames(patternRes)[oPat][1]]])
plotGeneCount(rd, crv, counts, gene = rownames(patternRes)[oPat][1])
```

We find genes at the top that are also ranked as DE for the differentiated cell type. What is especially interesting are genes that have different expression patterns but no different expression at the differentiated cell type level. We therefore sort the genes according to the sum of square of their rank in increasing Wald statistics for the *patternTest* and their rank in decreasing Wald statistics for the *diffEndTest*.

```{r}
compare <- inner_join(patternRes %>% mutate(Gene = rownames(patternRes),
                                            pattern = waldStat) %>%
                                     select(Gene, pattern),
                      endRes %>% mutate(Gene = rownames(endRes),
                                        end = waldStat) %>%
                                 select(Gene, end),
                      by = c("Gene" = "Gene")) %>%
           mutate(transientScore = (min_rank(desc(end)))^2 +
                                   (dense_rank(pattern))^2)

ggplot(compare, aes(x = log(pattern), y = log(end))) +
  geom_point(aes(col = transientScore)) +
  labs(x = "patternTest Wald Statistic (log scale)",
       y = "diffEndTest Wald Statistic (log scale)") +
  scale_color_continuous(low = "yellow", high = "red") +
  theme_classic()
```

Or, we can visualize the expression in UMAP space of the top gene.

```{r}
topTransient <- (compare %>% arrange(desc(transientScore)))[1, "Gene"]
plotSmoothers(gamList[[topTransient]])
plotGeneCount(rd, crv, counts, gene = topTransient)
```

Interestingly, we recover the Irf8 gene in the top 5 genes according to that ranking.

```{r}
head(compare %>% arrange(desc(transientScore)) %>% select(Gene), n = 5)
```

We can also plot the Irf8 gene.

```{r}
plotSmoothers(gamList[["Irf8"]])
plotGeneCount(rd, crv, counts, gene = "Irf8")
```

## Early drivers of differentiation

Another question of interest is to find a list of genes that are differentially expressed around the separation of two or multiple lineages.
The function `earlyDETest` implements a statistical method to tests the null hypothesis of whether the smoothers are equal between two user-specified knots by building on the `patternTest`, but restricting itself to a particular location of the smoothers.
Again, the knots can be visualized with the `plotGeneCount` function. By selecting the region covering the first two knot points to test for differential patterns between the lineages, we check which genes are behaving differently around the bifurcation point.

```{r}
plotGeneCount(rd = rd, curve = crv, counts = counts, clusters = cl,
              models = gamList)

earlyDERes <- earlyDETest(gamList, knots = c(1, 2))
oEarly <- order(earlyDERes$waldStat, decreasing = TRUE)
head(rownames(earlyDERes)[oEarly])
plotSmoothers(gamList[[rownames(earlyDERes)[oEarly][2]]])
plotGeneCount(rd, crv, counts, gene = rownames(earlyDERes)[oEarly][2])
```


# Clustering of genes according to their expression pattern

tradeSeq provides the functionality to cluster genes according to their expression pattern along the lineages with the `clusterExpressionPatterns` function. A number of equally spaced points for every lineage are selected to perform the clustering, and the number of points can be selected with the `nPoints` argument. The `genes` argument specifies which genes you want to cluster (e.g., all genes with differential expression patterns). Here, we use 20 points along each lineage to cluster the first 40 genes in the dataset.
The clustering itself occurs by the `clusterExperiment` package [@Risso2018], hence the user may select any clustering algorithm that's built into that package, or custom clustering algorithms implemented by the user. For a list of built-in clustering algorithms within `clusterExperiment`, run `clusterExperiment::listBuiltInFunctions()` on the command line.

```{r}
library(clusterExperiment)
nPointsClus <- 20
clusPat <- clusterExpressionPatterns(gamList, nPoints = nPointsClus,
                                     genes = rownames(counts)[1:40])
clusterLabels <- primaryCluster(clusPat$rsec)
```

The clusters can be visualized using the normalized expression that the clustering is based upon.

```{r}
par(mfrow = c(2, 2), bty = "l")
cUniq <- unique(clusterLabels)
cUniq <- cUniq[!cUniq == -1] # remove unclustered genes
for (xx in cUniq) {
  cId <- which(clusterLabels == xx)
  plot(x = 1:nPointsClus, y = rep(range(clusPat$yhatScaled[cId, ]),
                                  nPointsClus / 2), type = "n",
       main = paste0("Cluster ", xx), xlab = "Pseudotime",
       ylab = "Normalized expression")
  for (ii in 1:length(cId)) {
    geneId <- rownames(clusPat$yhatScaled)[cId[ii]]
    yhatGene <- clusPat$yhatScaled[geneId, ]
    lines(x = 1:nPointsClus, y = yhatGene[1:nPointsClus], col = "orange", lwd = 2)
    lines(x = 1:nPointsClus, y = yhatGene[(nPointsClus + 1):(2 * nPointsClus)],
          col = "darkseagreen3", lwd = 2)
  }
}
```

```{r clean the file that was downloaded when knitting the vignette}
file.remove("./se_paul.rda")
```

# Cheatsheet 

To recapitulate the workflow, we have created a cheatsheet that users can refer to when deciding which tests to run.

```{r, echo = F}
ggdraw() + draw_image("cheatsheet_highRes.jpeg")
```

# Session

```{r}
sessionInfo()
```

# References
