<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Using Monocle as input to **tradeSeq** • tradeSeq</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Using Monocle as input to **tradeSeq**">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">tradeSeq</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.1.13</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/tradeSeq.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Monocle.html">Using Monocle as input to **tradeSeq**</a>
    </li>
    <li>
      <a href="../articles/fitGAM.html">Fitting the models and additional control of fitGAM in **tradeSeq**</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Using Monocle as input to <strong>tradeSeq</strong>
</h1>
                        <h4 class="author">Koen Van den Berge and Hector Roux de Bézieux</h4>
            
            <h4 class="date">02/20/2020</h4>
      
      
      <div class="hidden name"><code>Monocle.Rmd</code></div>

    </div>

    
    
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p><code>tradeSeq</code> is an <code>R</code> package that allows analysis of gene expression along trajectories. While it has been developed and applied to single-cell RNA-sequencing (scRNA-seq) data, its applicability extends beyond that, and also allows the analysis of, e.g., single-cell ATAC-seq data along trajectories or bulk RNA-seq time series datasets. For every gene in the dataset, <code>tradeSeq</code> fits a generalized additive model (GAM) by building on the <code>mgcv</code> R package. It then allows statistical inference on the GAM by assessing contrasts of the parameters of the fitted GAM model, aiding in interpreting complex datasets. All details about the <code>tradeSeq</code> model and statistical tests are described in our preprint <span class="citation">(Van den Berge et al. 2019)</span>.</p>
<p>In this vignette, we analyze a subset of the data from <span class="citation">(Paul et al. 2015)</span>. A <code>SingleCellExperiment</code> object of the data has been provided with the <a href="https://github.com/statOmics/tradeSeqpaper"><code>tradeSeq</code></a> package and can be retrieved as shown below. The data and UMAP reduced dimensions were derived from following the <a href="http://cole-trapnell-lab.github.io/monocle-release/monocle3/#tutorial-1-learning-trajectories-with-monocle-3">Monocle 3 vignette</a>.</p>
<p>The main <a href="https://bioconductor.org/packages/release/bioc/vignettes/tradeSeq/inst/doc/tradeSeq.html">vignette</a> focuses on using <code>tradeSeq</code> downsteam of <code>slingshot</code>. Here, we present how to use <code>tradeSeq</code> downstream of <code>monocle</code><span class="citation">(Qiu et al. 2017)</span>.</p>
</div>
<div id="load-data" class="section level1">
<h1 class="hasAnchor">
<a href="#load-data" class="anchor"></a>Load data</h1>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(tradeSeq)</span>
<span id="cb1-2"><a href="#cb1-2"></a><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(RColorBrewer)</span>
<span id="cb1-3"><a href="#cb1-3"></a><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(SingleCellExperiment)</span>
<span id="cb1-4"><a href="#cb1-4"></a><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(monocle)</span>
<span id="cb1-5"><a href="#cb1-5"></a><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(dplyr)</span>
<span id="cb1-6"><a href="#cb1-6"></a><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(tidyr)</span>
<span id="cb1-7"><a href="#cb1-7"></a></span>
<span id="cb1-8"><a href="#cb1-8"></a><span class="co"># For reproducibility</span></span>
<span id="cb1-9"><a href="#cb1-9"></a><span class="kw"><a href="https://rdrr.io/r/grDevices/palette.html">palette</a></span>(<span class="kw"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html">brewer.pal</a></span>(<span class="dv">8</span>, <span class="st">"Dark2"</span>))</span>
<span id="cb1-10"><a href="#cb1-10"></a><span class="kw"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(countMatrix, <span class="dt">package =</span> <span class="st">"tradeSeq"</span>)</span>
<span id="cb1-11"><a href="#cb1-11"></a>counts &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span>(countMatrix)</span>
<span id="cb1-12"><a href="#cb1-12"></a><span class="kw"><a href="https://rdrr.io/r/base/rm.html">rm</a></span>(countMatrix)</span>
<span id="cb1-13"><a href="#cb1-13"></a><span class="kw"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(celltype, <span class="dt">package =</span> <span class="st">"tradeSeq"</span>)</span></code></pre></div>
</div>
<div id="fit-trajectories-using-monocle" class="section level1">
<h1 class="hasAnchor">
<a href="#fit-trajectories-using-monocle" class="anchor"></a>Fit trajectories using Monocle</h1>
<p>We will fit developmental trajectories using the <code>monocle</code> package.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1"></a><span class="kw"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="dv">200</span>)</span>
<span id="cb2-2"><a href="#cb2-2"></a>df &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="dt">cells =</span> <span class="kw"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(counts), <span class="dt">cellType =</span> celltype)</span>
<span id="cb2-3"><a href="#cb2-3"></a><span class="kw"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(df) &lt;-<span class="st"> </span>df<span class="op">$</span>cells</span>
<span id="cb2-4"><a href="#cb2-4"></a>cds &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/monocle/man/newCellDataSet.html">newCellDataSet</a></span>(counts, <span class="dt">phenoData =</span> <span class="kw">new</span>(<span class="st">"AnnotatedDataFrame"</span>, <span class="dt">data =</span> df))</span>
<span id="cb2-5"><a href="#cb2-5"></a>cds &lt;-<span class="st"> </span><span class="kw">estimateSizeFactors</span>(cds)</span>
<span id="cb2-6"><a href="#cb2-6"></a>cds &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/monocle/man/reduceDimension.html">reduceDimension</a></span>(cds, <span class="dt">max_components =</span> <span class="dv">2</span>)</span></code></pre></div>
<pre><code>## Registered S3 methods overwritten by 'proxy':
##   method               from    
##   print.registry_field registry
##   print.registry_entry registry</code></pre>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1"></a>cds &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/monocle/man/orderCells.html">orderCells</a></span>(cds)</span>
<span id="cb4-2"><a href="#cb4-2"></a>cds &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/monocle/man/orderCells.html">orderCells</a></span>(cds, <span class="dt">root_state =</span> <span class="dv">5</span>)</span></code></pre></div>
</div>
<div id="fit-smoothers" class="section level1">
<h1 class="hasAnchor">
<a href="#fit-smoothers" class="anchor"></a>Fit Smoothers</h1>
<p>Only possible for DDRTree</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a>sce &lt;-<span class="st"> </span><span class="kw"><a href="../reference/fitGAM.html">fitGAM</a></span>(cds, <span class="dt">verbose =</span> <span class="ot">TRUE</span>)</span></code></pre></div>
<p>You can then analyse the <code>sce</code> object following the main <a href="https://bioconductor.org/packages/release/bioc/vignettes/tradeSeq/inst/doc/tradeSeq.html#within-lineage-comparisons">vignette</a>.</p>
</div>
<div id="monocle3" class="section level1">
<h1 class="hasAnchor">
<a href="#monocle3" class="anchor"></a>Monocle3</h1>
<p>As of now (02/2020), monocle3<span class="citation">(<span class="citeproc-not-found" data-reference-id="Cao"><strong>???</strong></span>)</span>, is still in its beta version. Therefore, we have no plan yet to include a S4 method for monocle3 while it is not on CRAN or Bioconductor and the format is still moving. However, we present below a way to use <code>tradeSeq</code> downstream of <code>monocle3</code> as of version ‘0.2’, for a fully connected graph. We follow the tutorial from the <a href="https://cole-trapnell-lab.github.io/monocle3/docs/trajectories/">monocle3 website</a>.</p>
<div id="constructing-the-trajectory" class="section level2">
<h2 class="hasAnchor">
<a href="#constructing-the-trajectory" class="anchor"></a>Constructing the trajectory</h2>
<p>You will need to install monocle3 from <a href="https://cole-trapnell-lab.github.io/monocle3/docs/installation/">here</a> before running the code below.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1"></a><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(monocle3)</span>
<span id="cb6-2"><a href="#cb6-2"></a>cds &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/monocle3/man/new_cell_data_set.html">new_cell_data_set</a></span>(counts, <span class="dt">cell_metadata =</span> df,</span>
<span id="cb6-3"><a href="#cb6-3"></a>                <span class="dt">gene_metadata =</span> <span class="kw"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="dt">gene_short_name =</span> <span class="kw"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(counts),</span>
<span id="cb6-4"><a href="#cb6-4"></a>                                           <span class="dt">row.names =</span> <span class="kw"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(counts)))</span>
<span id="cb6-5"><a href="#cb6-5"></a>cds &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/monocle3/man/preprocess_cds.html">preprocess_cds</a></span>(cds, <span class="dt">method =</span> <span class="st">"PCA"</span>)</span>
<span id="cb6-6"><a href="#cb6-6"></a>cds &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/monocle3/man/reduce_dimension.html">reduce_dimension</a></span>(cds, <span class="dt">preprocess_method =</span> <span class="st">"PCA"</span>,</span>
<span id="cb6-7"><a href="#cb6-7"></a>                        <span class="dt">reduction_method =</span> <span class="st">"UMAP"</span>)</span>
<span id="cb6-8"><a href="#cb6-8"></a><span class="kw"><a href="https://rdrr.io/pkg/monocle3/man/plot_cells.html">plot_cells</a></span>(cds, <span class="dt">label_groups_by_cluster =</span> <span class="ot">FALSE</span>, <span class="dt">cell_size =</span> <span class="dv">1</span>,</span>
<span id="cb6-9"><a href="#cb6-9"></a>           <span class="dt">color_cells_by =</span> <span class="st">"cellType"</span>)</span>
<span id="cb6-10"><a href="#cb6-10"></a>cds &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/monocle3/man/cluster_cells.html">cluster_cells</a></span>(cds, <span class="dt">reduction_method =</span> <span class="st">"UMAP"</span>)</span>
<span id="cb6-11"><a href="#cb6-11"></a><span class="kw"><a href="https://rdrr.io/pkg/monocle3/man/plot_cells.html">plot_cells</a></span>(cds, <span class="dt">color_cells_by =</span> <span class="st">"cluster"</span>, <span class="dt">cell_size =</span> <span class="dv">1</span>)</span>
<span id="cb6-12"><a href="#cb6-12"></a>cds &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/monocle3/man/learn_graph.html">learn_graph</a></span>(cds, <span class="dt">use_partition =</span> <span class="ot">FALSE</span>)</span>
<span id="cb6-13"><a href="#cb6-13"></a><span class="co"># Find root </span></span>
<span id="cb6-14"><a href="#cb6-14"></a>cell_ids &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/which.html">which</a></span>(celltype <span class="op">==</span><span class="st">  "Multipotent progenitors"</span>)</span>
<span id="cb6-15"><a href="#cb6-15"></a>closest_vertex &lt;-<span class="st"> </span></span>
<span id="cb6-16"><a href="#cb6-16"></a><span class="st">  </span>cds<span class="op">@</span>principal_graph_aux[[<span class="st">"UMAP"</span>]]<span class="op">$</span>pr_graph_cell_proj_closest_vertex</span>
<span id="cb6-17"><a href="#cb6-17"></a>closest_vertex &lt;-<span class="st"> </span>closest_vertex[<span class="kw"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(cds), <span class="dv">1</span>]</span>
<span id="cb6-18"><a href="#cb6-18"></a>closest_vertex &lt;-<span class="st"> </span>closest_vertex[cell_ids]</span>
<span id="cb6-19"><a href="#cb6-19"></a>closest_vertex &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="st">"Y_"</span>, closest_vertex)</span>
<span id="cb6-20"><a href="#cb6-20"></a>root &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/which.html">which</a></span>(igraph<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/igraph/man/degree.html">degree</a></span>(<span class="kw"><a href="https://rdrr.io/pkg/monocle3/man/principal_graph.html">principal_graph</a></span>(cds)[[<span class="st">"UMAP"</span>]]) <span class="op">==</span><span class="st"> </span><span class="dv">1</span>))</span>
<span id="cb6-21"><a href="#cb6-21"></a>root &lt;-<span class="st"> </span>root[root <span class="op">%in%</span><span class="st"> </span>closest_vertex]</span>
<span id="cb6-22"><a href="#cb6-22"></a>cds &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/monocle3/man/order_cells.html">order_cells</a></span>(cds, <span class="dt">root_pr_nodes =</span> root)</span>
<span id="cb6-23"><a href="#cb6-23"></a><span class="kw"><a href="https://rdrr.io/pkg/monocle3/man/plot_cells.html">plot_cells</a></span>(cds, <span class="dt">color_cells_by =</span> <span class="st">"pseudotime"</span>)</span></code></pre></div>
</div>
<div id="extracting-the-pseudotimes-and-cell-weights-for-tradeseq" class="section level2">
<h2 class="hasAnchor">
<a href="#extracting-the-pseudotimes-and-cell-weights-for-tradeseq" class="anchor"></a>Extracting the pseudotimes and cell weights for tradeSeq</h2>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1"></a>y_to_cells &lt;-<span class="st">  </span><span class="kw"><a href="https://rdrr.io/pkg/monocle3/man/principal_graph_aux.html">principal_graph_aux</a></span>(cds)<span class="op">$</span>UMAP<span class="op">$</span>pr_graph_cell_proj_closest_vertex <span class="op">%&gt;%</span></span>
<span id="cb7-2"><a href="#cb7-2"></a><span class="st">  </span><span class="kw"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span>() <span class="op">%&gt;%</span></span>
<span id="cb7-3"><a href="#cb7-3"></a><span class="st">  </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(<span class="dt">cells =</span> <span class="kw"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(.)) <span class="op">%&gt;%</span></span>
<span id="cb7-4"><a href="#cb7-4"></a><span class="st">  </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/select.html">rename</a></span>(<span class="st">"Y"</span> =<span class="st"> </span>V1)</span>
<span id="cb7-5"><a href="#cb7-5"></a>mst &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/monocle3/man/principal_graph.html">principal_graph</a></span>(cds)<span class="op">$</span>UMAP</span>
<span id="cb7-6"><a href="#cb7-6"></a>endpoints &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/which.html">which</a></span>(igraph<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/igraph/man/degree.html">degree</a></span>(mst) <span class="op">==</span><span class="st"> </span><span class="dv">1</span>))</span>
<span id="cb7-7"><a href="#cb7-7"></a>endpoints &lt;-<span class="st"> </span>endpoints[endpoints <span class="op">!=</span><span class="st"> </span>root]</span>
<span id="cb7-8"><a href="#cb7-8"></a>cellWeights &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(endpoints, <span class="cf">function</span>(endpoint) {</span>
<span id="cb7-9"><a href="#cb7-9"></a>  path &lt;-<span class="st"> </span>igraph<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/igraph/man/distances.html">shortest_paths</a></span>(mst, root, endpoint)<span class="op">$</span>vpath[[<span class="dv">1</span>]]</span>
<span id="cb7-10"><a href="#cb7-10"></a>  path &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/character.html">as.character</a></span>(path)</span>
<span id="cb7-11"><a href="#cb7-11"></a>  df &lt;-<span class="st"> </span>y_to_cells <span class="op">%&gt;%</span></span>
<span id="cb7-12"><a href="#cb7-12"></a><span class="st">    </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(Y <span class="op">%in%</span><span class="st"> </span>path)</span>
<span id="cb7-13"><a href="#cb7-13"></a>  df &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="dt">weights =</span> <span class="kw"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(cds) <span class="op">%in%</span><span class="st"> </span>df<span class="op">$</span>cells))</span>
<span id="cb7-14"><a href="#cb7-14"></a>  <span class="kw"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(df) &lt;-<span class="st"> </span>endpoint</span>
<span id="cb7-15"><a href="#cb7-15"></a>  <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span>(df)</span>
<span id="cb7-16"><a href="#cb7-16"></a>  }) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/bind.html">bind_cols</a></span>() </span>
<span id="cb7-17"><a href="#cb7-17"></a>pseudotime &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span>(cellWeights, <span class="cf">function</span>(w) <span class="kw"><a href="https://rdrr.io/pkg/monocle3/man/pseudotime.html">pseudotime</a></span>(cds))</span>
<span id="cb7-18"><a href="#cb7-18"></a>sce &lt;-<span class="st"> </span><span class="kw"><a href="../reference/fitGAM.html">fitGAM</a></span>(<span class="dt">counts =</span> counts, </span>
<span id="cb7-19"><a href="#cb7-19"></a>       <span class="dt">pseudotime =</span> pseudotime,</span>
<span id="cb7-20"><a href="#cb7-20"></a>       <span class="dt">cellWeights =</span> cellWeights)</span></code></pre></div>
<p>Then, the <code>sce</code> object can be analyzed following the <a href="https://bioconductor.org/packages/release/bioc/vignettes/tradeSeq/inst/doc/tradeSeq.html">main vignette</a>.</p>
</div>
</div>
<div id="session" class="section level1">
<h1 class="hasAnchor">
<a href="#session" class="anchor"></a>Session</h1>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1"></a><span class="kw"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span>()</span></code></pre></div>
<pre><code>## R version 3.6.1 (2019-07-05)
## Platform: x86_64-apple-darwin15.6.0 (64-bit)
## Running under: macOS Catalina 10.15.3
## 
## Matrix products: default
## BLAS:   /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRblas.0.dylib
## LAPACK: /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRlapack.dylib
## 
## locale:
## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
## 
## attached base packages:
##  [1] splines   stats4    parallel  stats     graphics  grDevices utils    
##  [8] datasets  methods   base     
## 
## other attached packages:
##  [1] tidyr_1.0.2                 dplyr_0.8.4                
##  [3] monocle_2.14.0              DDRTree_0.1.5              
##  [5] irlba_2.3.3                 VGAM_1.1-2                 
##  [7] ggplot2_3.2.1               Matrix_1.2-18              
##  [9] SingleCellExperiment_1.8.0  SummarizedExperiment_1.16.1
## [11] DelayedArray_0.12.2         BiocParallel_1.20.1        
## [13] matrixStats_0.55.0          GenomicRanges_1.38.0       
## [15] GenomeInfoDb_1.22.0         IRanges_2.20.2             
## [17] S4Vectors_0.24.3            RColorBrewer_1.1-2         
## [19] tradeSeq_1.1.13             bigmemory_4.5.36           
## [21] Biobase_2.46.0              BiocGenerics_0.32.0        
## [23] knitr_1.28                 
## 
## loaded via a namespace (and not attached):
##   [1] uuid_0.1-2              backports_1.1.5         NMF_0.22.0             
##   [4] plyr_1.8.5              igraph_1.2.4.2          lazyeval_0.2.2         
##   [7] densityClust_0.3        rncl_0.8.4              fastICA_1.2-2          
##  [10] gridBase_0.4-7          digest_0.6.25           foreach_1.4.8          
##  [13] htmltools_0.4.0         viridis_0.5.1           magrittr_1.5           
##  [16] memoise_1.1.0           cluster_2.1.0           doParallel_1.0.15      
##  [19] limma_3.42.2            annotate_1.64.0         docopt_0.6.1           
##  [22] pkgdown_1.4.1           prettyunits_1.1.1       princurve_2.1.4        
##  [25] colorspace_1.4-1        blob_1.2.1              ggrepel_0.8.1          
##  [28] xfun_0.12               sparsesvd_0.2           crayon_1.3.4           
##  [31] RCurl_1.98-1.1          bigmemory.sri_0.1.3     genefilter_1.68.0      
##  [34] phylobase_0.8.8         survival_3.1-8          iterators_1.0.12       
##  [37] ape_5.3                 glue_1.3.1              registry_0.5-1         
##  [40] gtable_0.3.0            zlibbioc_1.32.0         XVector_0.26.0         
##  [43] kernlab_0.9-29          Rhdf5lib_1.8.0          HDF5Array_1.14.2       
##  [46] scales_1.1.0            pheatmap_1.0.12         DBI_1.1.0              
##  [49] edgeR_3.28.1            rngtools_1.5            bibtex_0.4.2.2         
##  [52] Rcpp_1.0.3              viridisLite_0.3.0       xtable_1.8-4           
##  [55] progress_1.2.2          proxy_0.4-23            bit_1.1-15.2           
##  [58] httr_1.4.1              FNN_1.1.3               pkgconfig_2.0.3        
##  [61] XML_3.99-0.3            locfit_1.5-9.1          howmany_0.3-1          
##  [64] tidyselect_1.0.0        rlang_0.4.4             softImpute_1.4         
##  [67] reshape2_1.4.3          AnnotationDbi_1.48.0    munsell_0.5.0          
##  [70] tools_3.6.1             RSQLite_2.2.0           ade4_1.7-15            
##  [73] evaluate_0.14           stringr_1.4.0           yaml_2.2.1             
##  [76] bit64_0.9-7             fs_1.3.1                purrr_0.3.3            
##  [79] RANN_2.6.1              pbapply_1.4-2           nlme_3.1-144           
##  [82] slam_0.1-47             xml2_1.2.2              compiler_3.6.1         
##  [85] rstudioapi_0.11         tibble_2.1.3            RNeXML_2.4.2           
##  [88] stringi_1.4.6           desc_1.2.0              RSpectra_0.16-0        
##  [91] lattice_0.20-40         HSMMSingleCell_1.6.0    vctrs_0.2.3            
##  [94] pillar_1.4.3            lifecycle_0.1.0         combinat_0.0-8         
##  [97] zinbwave_1.8.0          bitops_1.0-6            R6_2.4.1               
## [100] gridExtra_2.3           codetools_0.2-16        MASS_7.3-51.5          
## [103] assertthat_0.2.1        rhdf5_2.30.1            pkgmaker_0.31          
## [106] rprojroot_1.3-2         withr_2.1.2             qlcMatrix_0.9.7        
## [109] GenomeInfoDbData_1.2.2  locfdr_1.1-8            mgcv_1.8-31            
## [112] hms_0.5.3               grid_3.6.1              rmarkdown_2.1          
## [115] slingshot_1.4.0         Rtsne_0.15              clusterExperiment_2.6.1</code></pre>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<div id="refs" class="references">
<div id="ref-Paul2015">
<p>Paul, Franziska, Ya’ara Arkin, Amir Giladi, DiegoÂ Adhemar Jaitin, Ephraim Kenigsberg, Hadas Keren-Shaul, Deborah Winter, et al. 2015. “Transcriptional Heterogeneity and Lineage Commitment in Myeloid Progenitors.” <em>Cell</em> 163 (7): 1663–77. <a href="https://doi.org/10.1016/J.CELL.2015.11.013">https://doi.org/10.1016/J.CELL.2015.11.013</a>.</p>
</div>
<div id="ref-Qiu2017">
<p>Qiu, Xiaojie, Qi Mao, Ying Tang, Li Wang, Raghav Chawla, Hannah A Pliner, and Cole Trapnell. 2017. “Reversed graph embedding resolves complex single-cell trajectories.” <em>Nature Methods</em> 14 (10): 979–82. <a href="https://doi.org/10.1038/nmeth.4402">https://doi.org/10.1038/nmeth.4402</a>.</p>
</div>
<div id="ref-VandenBerge2019a">
<p>Van den Berge, Koen, Hector Roux de Bézieux, Kelly Street, Wouter Saelens, Robrecht Cannoodt, Yvan Saeys, Sandrine Dudoit, and Lieven Clement. 2019. “Trajectory-based differential expression analysis for single-cell sequencing data.” <em>bioRxiv</em>, May, 623397. <a href="https://doi.org/10.1101/623397">https://doi.org/10.1101/623397</a>.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">

        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#introduction">Introduction</a></li>
      <li><a href="#load-data">Load data</a></li>
      <li><a href="#fit-trajectories-using-monocle">Fit trajectories using Monocle</a></li>
      <li><a href="#fit-smoothers">Fit Smoothers</a></li>
      <li>
<a href="#monocle3">Monocle3</a><ul class="nav nav-pills nav-stacked">
<li><a href="#constructing-the-trajectory">Constructing the trajectory</a></li>
      <li><a href="#extracting-the-pseudotimes-and-cell-weights-for-tradeseq">Extracting the pseudotimes and cell weights for tradeSeq</a></li>
      </ul>
</li>
      <li><a href="#session">Session</a></li>
      <li><a href="#references">References</a></li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by Koen Van den Berge, Hector Roux de Bezieux.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
