<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Using Monocle as input to **tradeSeq** • tradeSeq</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Using Monocle as input to **tradeSeq**">
<meta property="og:description" content="tradeSeq">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">tradeSeq</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.1.18</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/tradeSeq.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Monocle.html">Using Monocle as input to **tradeSeq**</a>
    </li>
    <li>
      <a href="../articles/fitGAM.html">Fitting the models and additional control of fitGAM in **tradeSeq**</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/statOmics/tradeSeq/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Using Monocle as input to <strong>tradeSeq</strong>
</h1>
                        <h4 class="author">Koen Van den Berge and Hector Roux de Bézieux</h4>
            
            <h4 class="date">02/20/2020</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/statOmics/tradeSeq/blob/master/vignettes/Monocle.Rmd"><code>vignettes/Monocle.Rmd</code></a></small>
      <div class="hidden name"><code>Monocle.Rmd</code></div>

    </div>

    
    
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p><code>tradeSeq</code> is an <code>R</code> package that allows analysis of gene expression along trajectories. While it has been developed and applied to single-cell RNA-sequencing (scRNA-seq) data, its applicability extends beyond that, and also allows the analysis of, e.g., single-cell ATAC-seq data along trajectories or bulk RNA-seq time series datasets. For every gene in the dataset, <code>tradeSeq</code> fits a generalized additive model (GAM) by building on the <code>mgcv</code> R package. It then allows statistical inference on the GAM by assessing contrasts of the parameters of the fitted GAM model, aiding in interpreting complex datasets. All details about the <code>tradeSeq</code> model and statistical tests are described in our preprint <span class="citation">(Van den Berge et al. 2019)</span>.</p>
<p>In this vignette, we analyze a subset of the data from <span class="citation">(Paul et al. 2015)</span>. A <code>SingleCellExperiment</code> object of the data has been provided with the <a href="https://github.com/statOmics/tradeSeqpaper"><code>tradeSeq</code></a> package and can be retrieved as shown below. The data and UMAP reduced dimensions were derived from following the <a href="http://cole-trapnell-lab.github.io/monocle-release/monocle3/#tutorial-1-learning-trajectories-with-monocle-3">Monocle 3 vignette</a>.</p>
<p>The <a href="https://bioconductor.org/packages/devel/bioc/vignettes/tradeSeq/inst/doc/tradeSeq.html">main vignette</a> focuses on using <code>tradeSeq</code> downsteam of <code>slingshot</code>. Here, we present how to use <code>tradeSeq</code> downstream of <code>monocle</code><span class="citation">(Qiu et al. 2017)</span>.</p>
</div>
<div id="load-data" class="section level1">
<h1 class="hasAnchor">
<a href="#load-data" class="anchor"></a>Load data</h1>
<div class="sourceCode" id="cb1"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">tradeSeq</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">RColorBrewer</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">SingleCellExperiment</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">monocle</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">dplyr</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">tidyr</span>)

<span class="co"># For reproducibility</span>
<span class="fu"><a href="https://rdrr.io/r/grDevices/palette.html">palette</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html">brewer.pal</a></span>(<span class="fl">8</span>, <span class="st">"Dark2"</span>))
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="no">countMatrix</span>, <span class="kw">package</span> <span class="kw">=</span> <span class="st">"tradeSeq"</span>)
<span class="no">counts</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span>(<span class="no">countMatrix</span>)
<span class="fu"><a href="https://rdrr.io/r/base/rm.html">rm</a></span>(<span class="no">countMatrix</span>)
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="no">celltype</span>, <span class="kw">package</span> <span class="kw">=</span> <span class="st">"tradeSeq"</span>)</pre></body></html></div>
</div>
<div id="fit-trajectories-using-monocle" class="section level1">
<h1 class="hasAnchor">
<a href="#fit-trajectories-using-monocle" class="anchor"></a>Fit trajectories using Monocle</h1>
<p>We will fit developmental trajectories using the <code>monocle</code> package.</p>
<div class="sourceCode" id="cb2"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="fl">200</span>)
<span class="no">pd</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="kw">cells</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">counts</span>), <span class="kw">cellType</span> <span class="kw">=</span> <span class="no">celltype</span>)
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="no">pd</span>) <span class="kw">&lt;-</span> <span class="no">pd</span>$<span class="no">cells</span>
<span class="no">fd</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="kw">gene_short_name</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="no">counts</span>))
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="no">fd</span>) <span class="kw">&lt;-</span> <span class="no">fd</span>$<span class="no">gene_short_name</span>
<span class="no">cds</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/monocle/man/newCellDataSet.html">newCellDataSet</a></span>(<span class="no">counts</span>, <span class="kw">phenoData</span> <span class="kw">=</span> <span class="fu">new</span>(<span class="st">"AnnotatedDataFrame"</span>, <span class="kw">data</span> <span class="kw">=</span> <span class="no">pd</span>),
                      <span class="kw">featureData</span> <span class="kw">=</span> <span class="fu">new</span>(<span class="st">"AnnotatedDataFrame"</span>, <span class="kw">data</span> <span class="kw">=</span> <span class="no">fd</span>))
<span class="no">cds</span> <span class="kw">&lt;-</span> <span class="fu">estimateSizeFactors</span>(<span class="no">cds</span>)
<span class="no">cds</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/monocle/man/reduceDimension.html">reduceDimension</a></span>(<span class="no">cds</span>, <span class="kw">max_components</span> <span class="kw">=</span> <span class="fl">2</span>)</pre></body></html></div>
<pre><code>## Registered S3 methods overwritten by 'proxy':
##   method               from    
##   print.registry_field registry
##   print.registry_entry registry</code></pre>
<div class="sourceCode" id="cb4"><html><body><pre class="r"><span class="no">cds</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/monocle/man/orderCells.html">orderCells</a></span>(<span class="no">cds</span>)
<span class="no">cds</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/monocle/man/orderCells.html">orderCells</a></span>(<span class="no">cds</span>, <span class="kw">root_state</span> <span class="kw">=</span> <span class="fl">5</span>)</pre></body></html></div>
</div>
<div id="fit-smoothers" class="section level1">
<h1 class="hasAnchor">
<a href="#fit-smoothers" class="anchor"></a>Fit Smoothers</h1>
<p>Only possible for DDRTree</p>
<div class="sourceCode" id="cb5"><html><body><pre class="r"><span class="no">sce</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/fitGAM.html">fitGAM</a></span>(<span class="no">cds</span>, <span class="kw">verbose</span> <span class="kw">=</span> <span class="fl">TRUE</span>)</pre></body></html></div>
<p>You can then analyse the <code>sce</code> object following the <a href="https://bioconductor.org/packages/devel/bioc/vignettes/tradeSeq/inst/doc/tradeSeq.html">main vignette</a>.</p>
</div>
<div id="monocle3" class="section level1">
<h1 class="hasAnchor">
<a href="#monocle3" class="anchor"></a>Monocle3</h1>
<p>As of now (02/2020), monocle3<span class="citation">(Cao et al. 2019)</span>, is still in its beta version. Therefore, we have no plan yet to include a S4 method for monocle3 while it is not on CRAN or Bioconductor and the format is still moving. However, we present below a way to use <code>tradeSeq</code> downstream of <code>monocle3</code> as of version ‘0.2’, for a fully connected graph. We follow the tutorial from the <a href="https://cole-trapnell-lab.github.io/monocle3/docs/trajectories/">monocle3 website</a>.</p>
<div id="constructing-the-trajectory" class="section level2">
<h2 class="hasAnchor">
<a href="#constructing-the-trajectory" class="anchor"></a>Constructing the trajectory</h2>
<p>You will need to install monocle3 from <a href="https://cole-trapnell-lab.github.io/monocle3/docs/installation/">here</a> before running the code below.</p>
<div class="sourceCode" id="cb6"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">monocle3</span>)
<span class="co"># Create a cell_data_set object</span>
<span class="no">cds</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/monocle3/man/new_cell_data_set.html">new_cell_data_set</a></span>(<span class="no">counts</span>, <span class="kw">cell_metadata</span> <span class="kw">=</span> <span class="no">df</span>,
                <span class="kw">gene_metadata</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="kw">gene_short_name</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="no">counts</span>),
                                           <span class="kw">row.names</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="no">counts</span>)))
<span class="co"># Run PCA then UMAP on the data</span>
<span class="no">cds</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/monocle3/man/preprocess_cds.html">preprocess_cds</a></span>(<span class="no">cds</span>, <span class="kw">method</span> <span class="kw">=</span> <span class="st">"PCA"</span>)
<span class="no">cds</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/monocle3/man/reduce_dimension.html">reduce_dimension</a></span>(<span class="no">cds</span>, <span class="kw">preprocess_method</span> <span class="kw">=</span> <span class="st">"PCA"</span>,
                        <span class="kw">reduction_method</span> <span class="kw">=</span> <span class="st">"UMAP"</span>)
<span class="co"># First display, coloring by the cell types from Paul et al</span>
<span class="fu"><a href="https://rdrr.io/pkg/monocle3/man/plot_cells.html">plot_cells</a></span>(<span class="no">cds</span>, <span class="kw">label_groups_by_cluster</span> <span class="kw">=</span> <span class="fl">FALSE</span>, <span class="kw">cell_size</span> <span class="kw">=</span> <span class="fl">1</span>,
           <span class="kw">color_cells_by</span> <span class="kw">=</span> <span class="st">"cellType"</span>)

<span class="co"># Running the clustering method. This is necessary to the construct the graph</span>
<span class="no">cds</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/monocle3/man/cluster_cells.html">cluster_cells</a></span>(<span class="no">cds</span>, <span class="kw">reduction_method</span> <span class="kw">=</span> <span class="st">"UMAP"</span>)
<span class="co"># Visualize the clusters</span>
<span class="fu"><a href="https://rdrr.io/pkg/monocle3/man/plot_cells.html">plot_cells</a></span>(<span class="no">cds</span>, <span class="kw">color_cells_by</span> <span class="kw">=</span> <span class="st">"cluster"</span>, <span class="kw">cell_size</span> <span class="kw">=</span> <span class="fl">1</span>)
<span class="co"># Construct the graph</span>
<span class="co"># Note that, for the rest of the code to run, the graph should be fully connected</span>
<span class="no">cds</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/monocle3/man/learn_graph.html">learn_graph</a></span>(<span class="no">cds</span>, <span class="kw">use_partition</span> <span class="kw">=</span> <span class="fl">FALSE</span>)

<span class="co"># Find the root of the graph, i.e the node that is closest to the</span>
<span class="co"># cell labelled Multipotent progenitors</span>
<span class="no">cell_ids</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span>(<span class="no">celltype</span> <span class="kw">==</span>  <span class="st">"Multipotent progenitors"</span>)
<span class="no">closest_vertex</span> <span class="kw">&lt;-</span>
  <span class="no">cds</span>@<span class="kw">principal_graph_aux</span><span class="kw">[[</span><span class="st">"UMAP"</span>]]$<span class="no">pr_graph_cell_proj_closest_vertex</span>
<span class="no">closest_vertex</span> <span class="kw">&lt;-</span> <span class="no">closest_vertex</span>[<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">cds</span>), <span class="fl">1</span>]
<span class="no">closest_vertex</span> <span class="kw">&lt;-</span> <span class="no">closest_vertex</span>[<span class="no">cell_ids</span>]
<span class="no">closest_vertex</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="st">"Y_"</span>, <span class="no">closest_vertex</span>)
<span class="no">root</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span>(<span class="kw pkg">igraph</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/igraph/man/degree.html">degree</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/monocle3/man/principal_graph.html">principal_graph</a></span>(<span class="no">cds</span>)<span class="kw">[[</span><span class="st">"UMAP"</span>]]) <span class="kw">==</span> <span class="fl">1</span>))
<span class="no">root</span> <span class="kw">&lt;-</span> <span class="no">root</span>[<span class="no">root</span> <span class="kw">%in%</span> <span class="no">closest_vertex</span>]
<span class="no">cds</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/monocle3/man/order_cells.html">order_cells</a></span>(<span class="no">cds</span>, <span class="kw">root_pr_nodes</span> <span class="kw">=</span> <span class="no">root</span>)
<span class="fu"><a href="https://rdrr.io/pkg/monocle3/man/plot_cells.html">plot_cells</a></span>(<span class="no">cds</span>, <span class="kw">color_cells_by</span> <span class="kw">=</span> <span class="st">"pseudotime"</span>)</pre></body></html></div>
</div>
<div id="extracting-the-pseudotimes-and-cell-weights-for-tradeseq" class="section level2">
<h2 class="hasAnchor">
<a href="#extracting-the-pseudotimes-and-cell-weights-for-tradeseq" class="anchor"></a>Extracting the pseudotimes and cell weights for tradeSeq</h2>
<div class="sourceCode" id="cb7"><html><body><pre class="r"><span class="no">y_to_cells</span> <span class="kw">&lt;-</span>  <span class="fu"><a href="https://rdrr.io/pkg/monocle3/man/principal_graph_aux.html">principal_graph_aux</a></span>(<span class="no">cds</span>)$<span class="no">UMAP</span>$<span class="no">pr_graph_cell_proj_closest_vertex</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span>() <span class="kw">%&gt;%</span>
  <span class="kw pkg">dplyr</span><span class="kw ns">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(<span class="kw">cells</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="no">.</span>)) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">rename</a></span>(<span class="st">"Y"</span> <span class="kw">=</span> <span class="no">V1</span>)
<span class="no">mst</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/monocle3/man/principal_graph.html">principal_graph</a></span>(<span class="no">cds</span>)$<span class="no">UMAP</span>
<span class="no">endpoints</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span>(<span class="kw pkg">igraph</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/igraph/man/degree.html">degree</a></span>(<span class="no">mst</span>) <span class="kw">==</span> <span class="fl">1</span>))
<span class="no">endpoints</span> <span class="kw">&lt;-</span> <span class="no">endpoints</span>[<span class="no">endpoints</span> <span class="kw">!=</span> <span class="no">root</span>]
<span class="no">cellWeights</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(<span class="no">endpoints</span>, <span class="kw">function</span>(<span class="no">endpoint</span>) {
  <span class="no">path</span> <span class="kw">&lt;-</span> <span class="kw pkg">igraph</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/igraph/man/distances.html">shortest_paths</a></span>(<span class="no">mst</span>, <span class="no">root</span>, <span class="no">endpoint</span>)$<span class="no">vpath</span><span class="kw">[[</span><span class="fl">1</span>]]
  <span class="no">path</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span>(<span class="no">path</span>)
  <span class="no">df</span> <span class="kw">&lt;-</span> <span class="no">y_to_cells</span> <span class="kw">%&gt;%</span>
    <span class="kw pkg">dplyr</span><span class="kw ns">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(<span class="no">Y</span> <span class="kw">%in%</span> <span class="no">path</span>)
  <span class="no">df</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="kw">weights</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">cds</span>) <span class="kw">%in%</span> <span class="no">df</span>$<span class="no">cells</span>))
  <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">df</span>) <span class="kw">&lt;-</span> <span class="no">endpoint</span>
  <span class="fu"><a href="https://rdrr.io/r/base/function.html">return</a></span>(<span class="no">df</span>)
  }) <span class="kw">%&gt;%</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind.html">bind_cols</a></span>()
<span class="no">pseudotime</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span>(<span class="no">cellWeights</span>, <span class="kw">function</span>(<span class="no">w</span>) <span class="fu"><a href="https://rdrr.io/pkg/monocle3/man/pseudotime.html">pseudotime</a></span>(<span class="no">cds</span>))
<span class="no">sce</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/fitGAM.html">fitGAM</a></span>(<span class="kw">counts</span> <span class="kw">=</span> <span class="no">counts</span>,
       <span class="kw">pseudotime</span> <span class="kw">=</span> <span class="no">pseudotime</span>,
       <span class="kw">cellWeights</span> <span class="kw">=</span> <span class="no">cellWeights</span>)</pre></body></html></div>
<p>Then, the <code>sce</code> object can be analyzed following the <a href="https://bioconductor.org/packages/devel/bioc/vignettes/tradeSeq/inst/doc/tradeSeq.html">main vignette</a>.</p>
</div>
</div>
<div id="session" class="section level1">
<h1 class="hasAnchor">
<a href="#session" class="anchor"></a>Session</h1>
<div class="sourceCode" id="cb8"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span>()</pre></body></html></div>
<pre><code>## R version 3.6.1 (2019-07-05)
## Platform: x86_64-apple-darwin15.6.0 (64-bit)
## Running under: macOS Catalina 10.15.4
## 
## Matrix products: default
## BLAS:   /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRblas.0.dylib
## LAPACK: /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRlapack.dylib
## 
## locale:
## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
## 
## attached base packages:
##  [1] splines   stats4    parallel  stats     graphics  grDevices utils    
##  [8] datasets  methods   base     
## 
## other attached packages:
##  [1] tidyr_1.0.2                 dplyr_0.8.5                
##  [3] monocle_2.14.0              DDRTree_0.1.5              
##  [5] irlba_2.3.3                 VGAM_1.1-2                 
##  [7] ggplot2_3.3.0               Matrix_1.2-18              
##  [9] SingleCellExperiment_1.8.0  SummarizedExperiment_1.16.1
## [11] DelayedArray_0.12.2         BiocParallel_1.20.1        
## [13] matrixStats_0.56.0          GenomicRanges_1.38.0       
## [15] GenomeInfoDb_1.22.1         IRanges_2.20.2             
## [17] S4Vectors_0.24.3            RColorBrewer_1.1-2         
## [19] tradeSeq_1.1.18             bigmemory_4.5.36           
## [21] Biobase_2.46.0              BiocGenerics_0.32.0        
## [23] knitr_1.28                 
## 
## loaded via a namespace (and not attached):
##   [1] uuid_0.1-4              backports_1.1.6         NMF_0.22.0             
##   [4] plyr_1.8.6              igraph_1.2.5            lazyeval_0.2.2         
##   [7] densityClust_0.3        rncl_0.8.4              fastICA_1.2-2          
##  [10] gridBase_0.4-7          digest_0.6.25           foreach_1.5.0          
##  [13] htmltools_0.4.0         viridis_0.5.1           fansi_0.4.1            
##  [16] magrittr_1.5            memoise_1.1.0           cluster_2.1.0          
##  [19] doParallel_1.0.15       limma_3.42.2            annotate_1.64.0        
##  [22] docopt_0.6.1            pkgdown_1.5.0           prettyunits_1.1.1      
##  [25] princurve_2.1.4         colorspace_1.4-1        blob_1.2.1             
##  [28] ggrepel_0.8.2           xfun_0.12               sparsesvd_0.2          
##  [31] crayon_1.3.4            RCurl_1.98-1.1          bigmemory.sri_0.1.3    
##  [34] genefilter_1.68.0       phylobase_0.8.10        survival_3.1-11        
##  [37] iterators_1.0.12        ape_5.3                 glue_1.4.0             
##  [40] registry_0.5-1          gtable_0.3.0            zlibbioc_1.32.0        
##  [43] XVector_0.26.0          kernlab_0.9-29          Rhdf5lib_1.8.0         
##  [46] HDF5Array_1.14.3        scales_1.1.0            pheatmap_1.0.12        
##  [49] DBI_1.1.0               edgeR_3.28.1            rngtools_1.5           
##  [52] bibtex_0.4.2.2          Rcpp_1.0.4              viridisLite_0.3.0      
##  [55] xtable_1.8-4            progress_1.2.2          proxy_0.4-23           
##  [58] bit_1.1-15.2            httr_1.4.1              FNN_1.1.3              
##  [61] ellipsis_0.3.0          pkgconfig_2.0.3         XML_3.99-0.3           
##  [64] locfit_1.5-9.4          softImpute_1.4          howmany_0.3-1          
##  [67] tidyselect_1.0.0        rlang_0.4.5             reshape2_1.4.3         
##  [70] AnnotationDbi_1.48.0    munsell_0.5.0           tools_3.6.1            
##  [73] cli_2.0.2               RSQLite_2.2.0           ade4_1.7-15            
##  [76] evaluate_0.14           stringr_1.4.0           yaml_2.2.1             
##  [79] bit64_0.9-7             fs_1.4.0                purrr_0.3.3            
##  [82] RANN_2.6.1              pbapply_1.4-2           nlme_3.1-145           
##  [85] slam_0.1-47             xml2_1.3.0              compiler_3.6.1         
##  [88] rstudioapi_0.11         tibble_3.0.0            RNeXML_2.4.3           
##  [91] stringi_1.4.6           desc_1.2.0              RSpectra_0.16-0        
##  [94] lattice_0.20-40         HSMMSingleCell_1.6.0    vctrs_0.2.4            
##  [97] pillar_1.4.3            lifecycle_0.2.0         combinat_0.0-8         
## [100] zinbwave_1.8.0          bitops_1.0-6            R6_2.4.1               
## [103] gridExtra_2.3           codetools_0.2-16        MASS_7.3-51.5          
## [106] assertthat_0.2.1        rhdf5_2.30.1            pkgmaker_0.31.1        
## [109] rprojroot_1.3-2         withr_2.1.2             qlcMatrix_0.9.7        
## [112] GenomeInfoDbData_1.2.2  locfdr_1.1-8            mgcv_1.8-31            
## [115] hms_0.5.3               grid_3.6.1              rmarkdown_2.1          
## [118] slingshot_1.4.0         Rtsne_0.15              clusterExperiment_2.6.1</code></pre>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<div id="refs" class="references">
<div id="ref-Cao2019">
<p>Cao, Junyue, Malte Spielmann, Xiaojie Qiu, Xingfan Huang, Daniel M. Ibrahim, Andrew J. Hill, Fan Zhang, et al. 2019. “The Dynamics and Regulators of Cell Fate Decisions Are Revealed by Pseudo-Temporal Ordering of Single Cells.” <em>Nature</em>.</p>
</div>
<div id="ref-Paul2015">
<p>Paul, Franziska, Ya’ara Arkin, Amir Giladi, DiegoÂ Adhemar Jaitin, Ephraim Kenigsberg, Hadas Keren-Shaul, Deborah Winter, et al. 2015. “Transcriptional Heterogeneity and Lineage Commitment in Myeloid Progenitors.” <em>Cell</em> 163 (7): 1663–77. <a href="https://doi.org/10.1016/J.CELL.2015.11.013">https://doi.org/10.1016/J.CELL.2015.11.013</a>.</p>
</div>
<div id="ref-Qiu2017">
<p>Qiu, Xiaojie, Qi Mao, Ying Tang, Li Wang, Raghav Chawla, Hannah A Pliner, and Cole Trapnell. 2017. “Reversed graph embedding resolves complex single-cell trajectories.” <em>Nature Methods</em> 14 (10): 979–82. <a href="https://doi.org/10.1038/nmeth.4402">https://doi.org/10.1038/nmeth.4402</a>.</p>
</div>
<div id="ref-VandenBerge2019a">
<p>Van den Berge, Koen, Hector Roux de Bézieux, Kelly Street, Wouter Saelens, Robrecht Cannoodt, Yvan Saeys, Sandrine Dudoit, and Lieven Clement. 2019. “Trajectory-based differential expression analysis for single-cell sequencing data.” <em>bioRxiv</em>, May, 623397. <a href="https://doi.org/10.1101/623397">https://doi.org/10.1101/623397</a>.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Koen Van den Berge, Hector Roux de Bezieux.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.0.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
