<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Vignette for <strong>tradeSeq</strong>: <strong>tra</strong>jectory <strong>d</strong>ifferential <strong>e</strong>xpression analysis for <strong>Seq</strong>uncing data • tradeSeq</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Vignette for &lt;strong&gt;tradeSeq&lt;/strong&gt;: &lt;strong&gt;tra&lt;/strong&gt;jectory &lt;strong&gt;d&lt;/strong&gt;ifferential &lt;strong&gt;e&lt;/strong&gt;xpression analysis for &lt;strong&gt;Seq&lt;/strong&gt;uncing data">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">tradeSeq</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.9.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/tradeSeq.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Vignette for <strong>tradeSeq</strong>: <strong>tra</strong>jectory <strong>d</strong>ifferential <strong>e</strong>xpression analysis for <strong>Seq</strong>uncing data</h1>
                        <h4 class="author">Koen Van den Berge and Hector Roux de Bézieux</h4>
            
            <h4 class="date">9/10/2018</h4>
      
      
      <div class="hidden name"><code>tradeSeq.Rmd</code></div>

    </div>

    
    
<p>In this vignette, we analyse a subset of the data from <span class="citation">(Paul et al. 2015)</span>. A <code>SingleCellExperiment</code> object of the data has been provided with the <a href="https://github.com/statOmics/tradeR"><code>tradeSeq</code></a> package and can be retrieved with <code><a href="https://www.rdocumentation.org/packages/utils/topics/data">data(se)</a></code>. The data and UMAP reduced dimensions were derived from following the <a href="http://cole-trapnell-lab.github.io/monocle-release/monocle3/#tutorial-1-learning-trajectories-with-monocle-3">Monocle 3 vignette</a>.</p>
<div id="load-data" class="section level1">
<h1 class="hasAnchor">
<a href="#load-data" class="anchor"></a>Load data</h1>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/message">suppressPackageStartupMessages</a></span>({
  <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(SingleCellExperiment)
  <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(mgcv)
  <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(tradeSeq)
  <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(slingshot)
  <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(RColorBrewer)
  <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(dplyr)
  <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(ggplot2)
  <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(tidyr)
  <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(cowplot)
})
<span class="kw"><a href="https://www.rdocumentation.org/packages/grDevices/topics/palette">palette</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/RColorBrewer/topics/ColorBrewer">brewer.pal</a></span>(<span class="dv">8</span>,<span class="st">"Dark2"</span>))
<span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/data">data</a></span>(se, <span class="dt">package =</span> <span class="st">"tradeSeq"</span>)
se</code></pre></div>
<pre><code>## class: SingleCellExperiment 
## dim: 3004 2660 
## metadata(0):
## assays(1): counts
## rownames(3004): 0610007L01Rik 0610009O20Rik ... Zzef1 rp9
## rowData names(1): gene_short_name
## colnames(2660): W31105 W31106 ... W39167 W39168
## colData names(22): Seq_batch_ID Amp_batch_ID ... no_expression
##   louvain_component
## reducedDimNames(1): UMAP
## spikeNames(0):</code></pre>
</div>
<div id="fit-trajectories-using-slingshot" class="section level1">
<h1 class="hasAnchor">
<a href="#fit-trajectories-using-slingshot" class="anchor"></a>Fit trajectories using slingshot</h1>
<p>We will fit developmental trajectories using the <code>slingshot</code> package <span class="citation">(Street et al. 2018)</span>. <code>slingshot</code> requires cluster labels as input, and fits trajectories in reduced dimension. We will use the reduced space calculated with the UMAP method, which is pre-calculated in the <code>se</code> object. We cluster the data using k-means with <span class="math inline">\(7\)</span> clusters. Since we know which cells are the progenitor cell type, we define the starting point for the trajectories as input for <code>slingshot</code>. Note that this argument is optional, and not required to run <code>slingshot</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Random">set.seed</a></span>(<span class="dv">97</span>)
rd &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/SingleCellExperiment/topics/SCE-reducedDims">reducedDims</a></span>(se)<span class="op">$</span>UMAP
cl &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/kmeans">kmeans</a></span>(rd, <span class="dt">centers =</span> <span class="dv">7</span>)<span class="op">$</span>cluster
<span class="kw"><a href="https://www.rdocumentation.org/packages/graphics/topics/plot">plot</a></span>(rd, <span class="dt">col =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/RColorBrewer/topics/ColorBrewer">brewer.pal</a></span>(<span class="dv">9</span>, <span class="st">"Set1"</span>)[cl], <span class="dt">pch =</span> <span class="dv">16</span>, <span class="dt">asp =</span> <span class="dv">1</span>,
     <span class="dt">cex =</span> <span class="dv">2</span><span class="op">/</span><span class="dv">3</span>)</code></pre></div>
<p><img src="tradeSeq_files/figure-html/unnamed-chunk-2-1.png" width="700"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(slingshot)
lin &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/slingshot/topics/getLineages">getLineages</a></span>(rd, <span class="dt">clusterLabels =</span> cl, <span class="dt">start.clus =</span> <span class="dv">4</span>)</code></pre></div>
<pre><code>## Using full covariance matrix</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">crv &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/slingshot/topics/getCurves">getCurves</a></span>(lin)</code></pre></div>
<p>We find two lineages for this dataset. The trajectory can be visualized using the <code>plotGeneCount</code> function, using either the cluster labels or cell type to color the cells.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/plotGeneCount.html">plotGeneCount</a></span>(<span class="dt">rd =</span> rd, <span class="dt">curve =</span> crv, <span class="dt">counts =</span> counts, <span class="dt">clusters =</span> cl)</code></pre></div>
<p><img src="tradeSeq_files/figure-html/unnamed-chunk-3-1.png" width="700"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">celltype &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/factor">factor</a></span>(<span class="kw">colData</span>(se)<span class="op">$</span>cell_type2)

<span class="kw"><a href="../reference/plotGeneCount.html">plotGeneCount</a></span>(<span class="dt">rd =</span> rd, <span class="dt">curve =</span> crv, <span class="dt">counts =</span> counts, <span class="dt">clusters =</span> celltype, 
              <span class="dt">title=</span><span class="st">"Colored by cell type"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/graphics/topics/legend">legend</a></span>(<span class="st">"topright"</span>, <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/levels">levels</a></span>(celltype),
       <span class="dt">col =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/RColorBrewer/topics/ColorBrewer">brewer.pal</a></span>(<span class="dv">9</span>, <span class="st">"Set1"</span>)[<span class="dv">1</span><span class="op">:</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/nlevels">nlevels</a></span>(celltype)],
       <span class="dt">pch =</span> <span class="dv">16</span>, <span class="dt">cex =</span> <span class="dv">1</span> <span class="op">/</span><span class="st"> </span><span class="dv">2</span>, <span class="dt">bty =</span> <span class="st">"n"</span>)</code></pre></div>
<p><img src="tradeSeq_files/figure-html/unnamed-chunk-3-2.png" width="700"></p>
</div>
<div id="fit-additive-models" class="section level1">
<h1 class="hasAnchor">
<a href="#fit-additive-models" class="anchor"></a>Fit additive models</h1>
<p>After estimating the trajectory, we can fit generalized additive models (GAMs) with the <code>tradeSeq</code> package. Internally, this package builds on the <code>mgcv</code> package by fitting additive models using the <code>gam</code> function. The core function from <code>tradeSeq</code>, <code>fitGAM</code>, will use cubic splines as basis functions, and it tries to ensure that every lineage will end at a knot point of a smoother. By default, we allow for <span class="math inline">\(10\)</span> knots for every lineage, but this can be changed with the <code>nknots</code> function. More knots will allow more flexibility, but also increase the risk of overfitting. By default, the GAM model estimates one smoother for every lineage using the negative binomial distribution. If you want to allow for other fixed effects (e.g. batch effects), then an additional model matrix can be provided with the <code>U</code> argument.</p>
<p>We use the effective library size, estimated with TMM <span class="citation">(M. D. Robinson and Oshlack 2010)</span>, as offset in the model. We allow for alternatives by allowing a user-defined offset with the <code>offset</code> argument.</p>
<p>This dataset consists of UMI counts, and we do not expect zero inflation to be a big problem. However, we also allow to fit zero inflated negative binomial (ZINB) GAMs by providing observation-level weights to <code>fitGAM</code> using the <code>weights</code> argument. The <code>weights</code> must correspond to the posterior probability that a count belongs to the count component of the ZINB distribution <span class="citation">(Van den Berge et al. 2018)</span>.</p>
<p>For the vignette, we fit smoothers for a filtered set of genes in the dataset, 239 genes in total. We also include the <em>Irf8</em> gene, since it is a known transcription factor involved in hematopoiesis.</p>
<p>The progress of the fitting can be trackes using a progress bar by setting the <code>verbose</code> argument to <code>TRUE</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">counts &lt;-<span class="st"> </span><span class="kw">assays</span>(se)<span class="op">$</span>counts <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/matrix">as.matrix</a></span>()
filt &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colSums">rowSums</a></span>(counts <span class="op">&gt;</span><span class="st"> </span><span class="dv">8</span>) <span class="op">&gt;</span><span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/nrow">ncol</a></span>(counts)<span class="op">/</span><span class="dv">100</span>
filt[<span class="st">"Irf8"</span>] &lt;-<span class="st"> </span><span class="ot">TRUE</span>
counts &lt;-<span class="st"> </span>counts[filt, ]
gamList &lt;-<span class="st"> </span><span class="kw"><a href="../reference/fitGAM.html">fitGAM</a></span>(<span class="dt">counts =</span> counts,
                  <span class="dt">pseudotime =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/slingshot/topics/slingPseudotime">slingPseudotime</a></span>(crv, <span class="dt">na =</span> <span class="ot">FALSE</span>),
                  <span class="dt">cellWeights =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/slingshot/topics/slingPseudotime">slingCurveWeights</a></span>(crv),
                  <span class="dt">verbose =</span> <span class="ot">FALSE</span>)
<span class="co"># This takes about 1mn to run</span></code></pre></div>
<p>One may explore the results of a model by requesting its summary.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/summary">summary</a></span>(gamList[[<span class="st">"Irf8"</span>]])</code></pre></div>
<pre><code>## 
## Family: Negative Binomial(2.32) 
## Link function: log 
## 
## Formula:
## y ~ -1 + U + s(t1, by = l1, bs = "cr", id = 1, k = nknots) + 
##     s(t2, by = l2, bs = "cr", id = 1, k = nknots) + offset(offset)
## 
## Parametric coefficients:
##   Estimate Std. Error z value Pr(&gt;|z|)    
## U -10.0039     0.5687  -17.59   &lt;2e-16 ***
## ---
## Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1
## 
## Approximate significance of smooth terms:
##            edf Ref.df  Chi.sq p-value    
## s(t1):l1 8.555  9.477 624.177  &lt;2e-16 ***
## s(t2):l2 4.403  5.240   6.489   0.261    
## ---
## Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1
## 
## Rank: 20/21
## R-sq.(adj) =   0.54   Deviance explained =   65%
## -REML = 2014.3  Scale est. = 1         n = 2660</code></pre>
<p>You can also plot the cells in reduced dimension to see where the knots are located.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/plotGeneCount.html">plotGeneCount</a></span>(<span class="dt">rd =</span> rd, <span class="dt">curve =</span> crv, <span class="dt">counts =</span> counts, <span class="dt">clusters =</span> cl,
              <span class="dt">models =</span> gamList)</code></pre></div>
<p><img src="tradeSeq_files/figure-html/unnamed-chunk-6-1.png" width="700"></p>
</div>
<div id="within-lineage-comparisons" class="section level1">
<h1 class="hasAnchor">
<a href="#within-lineage-comparisons" class="anchor"></a>Within-lineage comparisons</h1>
<div id="association-of-gene-expression-with-pseudotime" class="section level2">
<h2 class="hasAnchor">
<a href="#association-of-gene-expression-with-pseudotime" class="anchor"></a>Association of gene expression with pseudotime</h2>
<p>A first exploration of the data analysis may consist in checking whether gene expression is associated with a particular lineage. The statistical test performed here, implemented in the <code>associationTest</code> function, is testing the null hypothesis that all smoother coefficients are equal to each other. This can be interpreted as testing whether the smoothed gene expression is significantly changing along pseudotime.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">assoRes &lt;-<span class="st"> </span><span class="kw"><a href="../reference/associationTest.html">associationTest</a></span>(gamList)</code></pre></div>
<p>Related, one can extract the p-values generated by the <code>mgcv</code> package using the <code>getSmootherPvalues</code> function. These p-values are derived from a test that assesses the null hypothesis that all smoother coefficients are equal to zero. Note, however, that their interpretation is thus more complex. A significant lineage for a particular gene might thus be the result of (a) a different mean expression in that lineage as compared to the overall expression of that gene, or (b) significantly varying expression along that lineage, even if the means are equal, or (c) a combination of both. This function extracts the p-values calculated by <code>mgcv</code> from the GAM, and will return <code>NA</code> for genes that we were unable to fit properly. Similarly, the test statistics may be extracted with <code>getSmootherTestStats</code>. Since this dataset was pre-filtered to only contain relevant genes, all p-values (test statistics) will be very low (high).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pvalLineage &lt;-<span class="st"> </span><span class="kw"><a href="../reference/getSmootherPvalues.html">getSmootherPvalues</a></span>(gamList)
statLineage &lt;-<span class="st"> </span><span class="kw"><a href="../reference/getSmootherTestStats.html">getSmootherTestStats</a></span>(gamList)</code></pre></div>
</div>
<div id="discovering-progenitor-marker-genes" class="section level2">
<h2 class="hasAnchor">
<a href="#discovering-progenitor-marker-genes" class="anchor"></a>Discovering progenitor marker genes</h2>
<p>In order to discover marker genes of the progenitor cell population, researchers may be interested in assessing differential expression between the progenitor cell population (i.e., the starting point of a lineage) with the differentiated cell type population (i.e., the end point of a lineage). In the function <code>startVsEndTest</code>, we have implemented a Wald test that tests the null hypothesis that the expression at the starting point of the smoother (progenitor population) is identical to the expression at the end point of the smoother (differentiated population). The test basically involves a comparison between two smoother coefficients for every lineage. The function <code>startVsEndTest</code> performs an omnibus test across all lineages by default, but you can also assess all lienages separately by setting <code>lineages=TRUE</code>. Below, we adopt an omnibus test across the two lineages.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">startRes &lt;-<span class="st"> </span><span class="kw"><a href="../reference/startVsEndTest.html">startVsEndTest</a></span>(gamList)</code></pre></div>
<p>We can visualize the estimated smoothers for the most significant gene.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">oStart &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/order">order</a></span>(startRes<span class="op">$</span>waldStat, <span class="dt">decreasing =</span> <span class="ot">TRUE</span>)
sigGeneStart &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(gamList)[oStart[<span class="dv">1</span>]]
<span class="kw"><a href="../reference/plotSmoothers.html">plotSmoothers</a></span>(gamList[[sigGeneStart]])</code></pre></div>
<p><img src="tradeSeq_files/figure-html/unnamed-chunk-10-1.png" width="700"></p>
<p>Alternatively, we can color the cells in UMAP space with that gene’s expression.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/plotGeneCount.html">plotGeneCount</a></span>(rd, crv, counts, <span class="dt">gene =</span> sigGeneStart)</code></pre></div>
<p><img src="tradeSeq_files/figure-html/unnamed-chunk-11-1.png" width="700"></p>
</div>
</div>
<div id="between-lineage-comparisons" class="section level1">
<h1 class="hasAnchor">
<a href="#between-lineage-comparisons" class="anchor"></a>Between-lineage comparisons</h1>
<div id="discovering-differentiated-cell-type-markers" class="section level2">
<h2 class="hasAnchor">
<a href="#discovering-differentiated-cell-type-markers" class="anchor"></a>Discovering differentiated cell type markers</h2>
<p><code>tradeSeq</code> can discover marker genes for the differentiated cell types by comparing the end points of the lineage-specific smoothers. This is implemented in the <code>diffEndTest</code> function. By default, <code>diffEndTest</code> performs an omnibus test, testing the null hypothesis that the endpoint expression is equal for all lineages using a multivariate Wald test. If more than two trajectories are present, one can assess all pairwise comparisons using the <code>pairwise=TRUE</code> argument.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">endRes &lt;-<span class="st"> </span><span class="kw"><a href="../reference/diffEndTest.html">diffEndTest</a></span>(gamList)</code></pre></div>
<p>We can plot the most significant gene using the <code>plotSmoothers</code> function.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">o &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/order">order</a></span>(endRes<span class="op">$</span>waldStat, <span class="dt">decreasing =</span> <span class="ot">TRUE</span>)
sigGene &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(gamList)[o[<span class="dv">2</span>]]
<span class="kw"><a href="../reference/plotSmoothers.html">plotSmoothers</a></span>(gamList[[sigGene]])</code></pre></div>
<p><img src="tradeSeq_files/figure-html/unnamed-chunk-13-1.png" width="700"></p>
<p>Alternatively, we can color the cells in UMAP space with that gene’s expression.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/plotGeneCount.html">plotGeneCount</a></span>(rd, crv, counts, <span class="dt">gene =</span> sigGene)</code></pre></div>
<p><img src="tradeSeq_files/figure-html/unnamed-chunk-14-1.png" width="700"></p>
</div>
<div id="discovering-genes-with-different-expression-patterns" class="section level2">
<h2 class="hasAnchor">
<a href="#discovering-genes-with-different-expression-patterns" class="anchor"></a>Discovering genes with different expression patterns</h2>
<p>Asides from testing at the level of the differentiated cell type, researchers may be interested in assessing the expression pattern of a gene over pseudotime. The function <code>patternTest</code> implements a statistical method that checks whether the smoothed gene expression is equal along pseudotime between two or multiple lineages. In practice, we use <span class="math inline">\(100\)</span> points, equally distributed along pseudotime, that are compared between two (or multiple) lineages, and this number can be changed using the <code>nPoints</code> argument.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">patternRes &lt;-<span class="st"> </span><span class="kw"><a href="../reference/patternTest.html">patternTest</a></span>(gamList)
oPat &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/order">order</a></span>(patternRes<span class="op">$</span>waldStat, <span class="dt">decreasing =</span> <span class="ot">TRUE</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/head">head</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">rownames</a></span>(patternRes)[oPat])</code></pre></div>
<pre><code>## [1] "Mpo"   "Prtn3" "Car2"  "Ctsg"  "Elane" "H2afy"</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/plotSmoothers.html">plotSmoothers</a></span>(gamList[[<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">rownames</a></span>(patternRes)[oPat][<span class="dv">1</span>]]])</code></pre></div>
<p><img src="tradeSeq_files/figure-html/unnamed-chunk-15-1.png" width="700"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/plotGeneCount.html">plotGeneCount</a></span>(rd, crv, counts, <span class="dt">gene =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">rownames</a></span>(patternRes)[oPat][<span class="dv">1</span>])</code></pre></div>
<p><img src="tradeSeq_files/figure-html/unnamed-chunk-15-2.png" width="700"></p>
<p>We find genes at the top that are also ranked as DE for the differentiated cell type. What is especially interesting are genes that have different expression patterns but no different expression at the differentiated cell type level. We therefore sort the genes according to the sum of square of their rank in increasing Wald statistics for the <em>patternTest</em> and their rank in decreasing Wald statistics for the <em>diffEndTest</em>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">compare &lt;-<span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/join.html">inner_join</a></span>(patternRes <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(<span class="dt">Gene =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">rownames</a></span>(patternRes),
                                            <span class="dt">pattern =</span> waldStat) <span class="op">%&gt;%</span>
<span class="st">                                     </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span>(Gene, pattern),
                      endRes <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(<span class="dt">Gene =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">rownames</a></span>(endRes),
                                        <span class="dt">end =</span> waldStat) <span class="op">%&gt;%</span>
<span class="st">                                 </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span>(Gene, end),
                      <span class="dt">by =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"Gene"</span> =<span class="st"> "Gene"</span>)) <span class="op">%&gt;%</span>
<span class="st">           </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(<span class="dt">transientScore =</span> (<span class="kw"><a href="https://dplyr.tidyverse.org/reference/ranking.html">min_rank</a></span>(<span class="kw"><a href="https://dplyr.tidyverse.org/reference/desc.html">desc</a></span>(end)))<span class="op">^</span><span class="dv">2</span> <span class="op">+</span>
<span class="st">                                   </span>(<span class="kw"><a href="https://dplyr.tidyverse.org/reference/ranking.html">dense_rank</a></span>(pattern))<span class="op">^</span><span class="dv">2</span>)

<span class="kw"><a href="https://www.rdocumentation.org/packages/ggplot2/topics/ggplot">ggplot</a></span>(compare, <span class="kw"><a href="https://www.rdocumentation.org/packages/ggplot2/topics/aes">aes</a></span>(<span class="dt">x =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Log">log</a></span>(pattern), <span class="dt">y =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Log">log</a></span>(end))) <span class="op">+</span>
<span class="st">  </span><span class="kw"><a href="https://www.rdocumentation.org/packages/ggplot2/topics/geom_point">geom_point</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/ggplot2/topics/aes">aes</a></span>(<span class="dt">col =</span> transientScore)) <span class="op">+</span>
<span class="st">  </span><span class="kw"><a href="https://www.rdocumentation.org/packages/ggplot2/topics/labs">labs</a></span>(<span class="dt">x =</span> <span class="st">"patternTest Wald Statistic (log scale)"</span>,
       <span class="dt">y =</span> <span class="st">"diffEndTest Wald Statistic (log scale)"</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw"><a href="https://www.rdocumentation.org/packages/ggplot2/topics/scale_gradient">scale_color_continuous</a></span>(<span class="dt">low =</span> <span class="st">"yellow"</span>, <span class="dt">high =</span> <span class="st">"red"</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw"><a href="https://www.rdocumentation.org/packages/ggplot2/topics/ggtheme">theme_classic</a></span>()</code></pre></div>
<p><img src="tradeSeq_files/figure-html/unnamed-chunk-16-1.png" width="700"></p>
<p>Or, we can visualize the expression in UMAP space of the top gene.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">topTransient &lt;-<span class="st"> </span>(compare <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span>(<span class="kw"><a href="https://dplyr.tidyverse.org/reference/desc.html">desc</a></span>(transientScore)))[<span class="dv">1</span>, <span class="st">"Gene"</span>]
<span class="kw"><a href="../reference/plotSmoothers.html">plotSmoothers</a></span>(gamList[[topTransient]])</code></pre></div>
<p><img src="tradeSeq_files/figure-html/unnamed-chunk-17-1.png" width="700"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/plotGeneCount.html">plotGeneCount</a></span>(rd, crv, counts, <span class="dt">gene =</span> topTransient)</code></pre></div>
<p><img src="tradeSeq_files/figure-html/unnamed-chunk-17-2.png" width="700"></p>
<p>Interestingly, we recover the Irf8 gene in the top 5 genes according to that ranking.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/head">head</a></span>(compare <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span>(<span class="kw"><a href="https://dplyr.tidyverse.org/reference/desc.html">desc</a></span>(transientScore)) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span>(Gene), <span class="dt">n =</span> <span class="dv">5</span>)</code></pre></div>
<pre><code>##    Gene
## 1 Nedd4
## 2   Mt1
## 3 Mki67
## 4  Irf8
## 5 Hint1</code></pre>
<p>We can also plot the Irf8 gene.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/plotSmoothers.html">plotSmoothers</a></span>(gamList[[<span class="st">"Irf8"</span>]])</code></pre></div>
<p><img src="tradeSeq_files/figure-html/unnamed-chunk-19-1.png" width="700"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/plotGeneCount.html">plotGeneCount</a></span>(rd, crv, counts, <span class="dt">gene =</span> <span class="st">"Irf8"</span>)</code></pre></div>
<p><img src="tradeSeq_files/figure-html/unnamed-chunk-19-2.png" width="700"></p>
</div>
<div id="early-drivers-of-differentiation" class="section level2">
<h2 class="hasAnchor">
<a href="#early-drivers-of-differentiation" class="anchor"></a>Early drivers of differentiation</h2>
<p>Another question of interest is to find a list of genes that are differentially expressed around the separation of two or multiple lineages. The function <code>earlyDETest</code> implements a statistical method to tests the null hypothesis of whether the smoothers are equal between two user-specified knots by building on the <code>patternTest</code>, but restricting itself to a particular location of the smoothers. Again, the knots can be visualized with the <code>plotGeneCount</code> function. By selecting the region covering the first two knot points to test for differential patterns between the lineages, we check which genes are behaving differently around the bifurcation point.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/plotGeneCount.html">plotGeneCount</a></span>(<span class="dt">rd =</span> rd, <span class="dt">curve =</span> crv, <span class="dt">counts =</span> counts, <span class="dt">clusters =</span> cl,
              <span class="dt">models =</span> gamList)</code></pre></div>
<p><img src="tradeSeq_files/figure-html/unnamed-chunk-20-1.png" width="700"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">earlyDERes &lt;-<span class="st"> </span><span class="kw"><a href="../reference/earlyDETest.html">earlyDETest</a></span>(gamList, <span class="dt">knots =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">1</span>, <span class="dv">2</span>))
oEarly &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/order">order</a></span>(earlyDERes<span class="op">$</span>waldStat, <span class="dt">decreasing =</span> <span class="ot">TRUE</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/head">head</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">rownames</a></span>(earlyDERes)[oEarly])</code></pre></div>
<pre><code>## [1] "Alas1" "Car1"  "Car2"  "Mt1"   "Klf1"  "Pkm2"</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/plotSmoothers.html">plotSmoothers</a></span>(gamList[[<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">rownames</a></span>(earlyDERes)[oEarly][<span class="dv">2</span>]]])</code></pre></div>
<p><img src="tradeSeq_files/figure-html/unnamed-chunk-20-2.png" width="700"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/plotGeneCount.html">plotGeneCount</a></span>(rd, crv, counts, <span class="dt">gene =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">rownames</a></span>(earlyDERes)[oEarly][<span class="dv">2</span>])</code></pre></div>
<p><img src="tradeSeq_files/figure-html/unnamed-chunk-20-3.png" width="700"></p>
</div>
</div>
<div id="clustering-of-genes-according-to-their-expression-pattern" class="section level1">
<h1 class="hasAnchor">
<a href="#clustering-of-genes-according-to-their-expression-pattern" class="anchor"></a>Clustering of genes according to their expression pattern</h1>
<p>tradeSeq provides the functionality to cluster genes according to their expression pattern along the lineages with the <code>clusterExpressionPatterns</code> function. A number of equally spaced points for every lineage are selected to perform the clustering, and the number of points can be selected with the <code>nPoints</code> argument. The <code>genes</code> argument specifies which genes you want to cluster (e.g., all genes with differential expression patterns). Here, we use 20 points along each lineage to cluster the first 40 genes in the dataset. The clustering itself occurs with the RSEC package <span class="citation">(Risso et al. 2018)</span>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(clusterExperiment)</code></pre></div>
<pre><code>## Warning: package 'clusterExperiment' was built under R version 3.5.1</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">nPointsClus &lt;-<span class="st"> </span><span class="dv">20</span>
clusPat &lt;-<span class="st"> </span><span class="kw"><a href="../reference/clusterExpressionPatterns.html">clusterExpressionPatterns</a></span>(gamList, <span class="dt">nPoints=</span>nPointsClus, <span class="dt">genes=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">rownames</a></span>(counts)[<span class="dv">1</span><span class="op">:</span><span class="dv">40</span>])</code></pre></div>
<pre><code>## Note: Not all of the methods requested in 'reduceMethod' have been calculated. Will calculate all the methods requested (any pre-existing values -- filtering statistics or dimensionality reductions -- with these names will be recalculated and overwritten): PCA.</code></pre>
<pre><code>## Note: Merging will be done on ' makeConsensus ', with clustering index 1</code></pre>
<pre><code>## Note: merging with these parameters did not result in any clusters being merged.</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">clusterLabels &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/clusterExperiment/topics/ClusterExperiment-methods">primaryCluster</a></span>(clusPat<span class="op">$</span>rsec)</code></pre></div>
<p>The clusters can be visualized using the normalized expression that the clustering is based upon.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/graphics/topics/par">par</a></span>(<span class="dt">mfrow=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">2</span>,<span class="dv">2</span>), <span class="dt">bty=</span><span class="st">'l'</span>)
cUniq &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/unique">unique</a></span>(clusterLabels)
cUniq &lt;-<span class="st"> </span>cUniq[<span class="op">!</span>cUniq<span class="op">==-</span><span class="dv">1</span>] <span class="co">#remove unclustered genes</span>
<span class="cf">for</span>(xx <span class="cf">in</span> cUniq){
  cId &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/which">which</a></span>(clusterLabels<span class="op">==</span>xx)
  <span class="kw"><a href="https://www.rdocumentation.org/packages/graphics/topics/plot">plot</a></span>(<span class="dt">x=</span><span class="dv">1</span><span class="op">:</span>nPointsClus,<span class="dt">y=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/rep">rep</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/range">range</a></span>(clusPat<span class="op">$</span>yhatScaled[cId,]),nPointsClus<span class="op">/</span><span class="dv">2</span>), <span class="dt">type=</span><span class="st">"n"</span>, <span class="dt">main=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste0</a></span>(<span class="st">"Cluster "</span>,xx), <span class="dt">xlab=</span><span class="st">"Pseudotime"</span>, <span class="dt">ylab=</span><span class="st">"Normalized expression"</span>)
  <span class="cf">for</span>(ii <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/length">length</a></span>(cId)){
    geneId &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">rownames</a></span>(clusPat<span class="op">$</span>yhatScaled)[cId[ii]]
    yhatGene &lt;-<span class="st"> </span>clusPat<span class="op">$</span>yhatScaled[geneId,]
    <span class="kw"><a href="https://www.rdocumentation.org/packages/graphics/topics/lines">lines</a></span>(<span class="dt">x=</span><span class="dv">1</span><span class="op">:</span>nPointsClus, <span class="dt">y=</span>yhatGene[<span class="dv">1</span><span class="op">:</span>nPointsClus], <span class="dt">col=</span><span class="st">"orange"</span>, <span class="dt">lwd=</span><span class="dv">2</span>)
    <span class="kw"><a href="https://www.rdocumentation.org/packages/graphics/topics/lines">lines</a></span>(<span class="dt">x=</span><span class="dv">1</span><span class="op">:</span>nPointsClus, <span class="dt">y=</span>yhatGene[(nPointsClus<span class="op">+</span><span class="dv">1</span>)<span class="op">:</span>(<span class="dv">2</span><span class="op">*</span>nPointsClus)], <span class="dt">col=</span><span class="st">"darkseagreen3"</span>, <span class="dt">lwd=</span><span class="dv">2</span>)
  }
}</code></pre></div>
<p><img src="tradeSeq_files/figure-html/unnamed-chunk-22-1.png" width="700"></p>
</div>
<div id="cheatsheet" class="section level1">
<h1 class="hasAnchor">
<a href="#cheatsheet" class="anchor"></a>Cheatsheet</h1>
<p>To recapitulate the workflow, we have created a cheatsheet that users can refer to when deciding which tests to run.</p>
<p><img src="tradeSeq_files/figure-html/unnamed-chunk-23-1.png" width="700"></p>
</div>
<div id="session" class="section level1">
<h1 class="hasAnchor">
<a href="#session" class="anchor"></a>Session</h1>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/sessionInfo">sessionInfo</a></span>()</code></pre></div>
<pre><code>## R version 3.5.0 (2018-04-23)
## Platform: x86_64-apple-darwin15.6.0 (64-bit)
## Running under: macOS  10.14.3
## 
## Matrix products: default
## BLAS: /Library/Frameworks/R.framework/Versions/3.5/Resources/lib/libRblas.0.dylib
## LAPACK: /Library/Frameworks/R.framework/Versions/3.5/Resources/lib/libRlapack.dylib
## 
## locale:
## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
## 
## attached base packages:
## [1] parallel  stats4    stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] clusterExperiment_2.2.0     cowplot_0.9.4              
##  [3] tidyr_0.8.3                 ggplot2_3.1.0              
##  [5] dplyr_0.8.0.1               RColorBrewer_1.1-2         
##  [7] slingshot_1.0.0             princurve_2.1.3            
##  [9] tradeSeq_0.9.0              bigmemory_4.5.33           
## [11] mgcv_1.8-27                 nlme_3.1-137               
## [13] SingleCellExperiment_1.4.1  SummarizedExperiment_1.12.0
## [15] DelayedArray_0.8.0          BiocParallel_1.16.6        
## [17] matrixStats_0.54.0          Biobase_2.42.0             
## [19] GenomicRanges_1.34.0        GenomeInfoDb_1.18.2        
## [21] IRanges_2.16.0              S4Vectors_0.20.1           
## [23] BiocGenerics_0.28.0        
## 
## loaded via a namespace (and not attached):
##   [1] uuid_0.1-2              backports_1.1.3        
##   [3] copula_0.999-19         NMF_0.21.0             
##   [5] igraph_1.2.4            plyr_1.8.4             
##   [7] lazyeval_0.2.1          splines_3.5.0          
##   [9] pspline_1.0-18          crosstalk_1.0.0        
##  [11] rncl_0.8.3              gridBase_0.4-7         
##  [13] digest_0.6.18           foreach_1.4.4          
##  [15] htmltools_0.3.6         magick_2.0             
##  [17] viridis_0.5.1           magrittr_1.5           
##  [19] memoise_1.1.0           cluster_2.0.7-1        
##  [21] doParallel_1.0.14       limma_3.38.3           
##  [23] annotate_1.60.1         stabledist_0.7-1       
##  [25] pkgdown_1.3.0           prettyunits_1.0.2      
##  [27] colorspace_1.4-0        blob_1.1.1             
##  [29] xfun_0.5                jsonlite_1.6           
##  [31] crayon_1.3.4            RCurl_1.95-4.12        
##  [33] bigmemory.sri_0.1.3     roxygen2_6.1.1         
##  [35] genefilter_1.64.0       phylobase_0.8.6        
##  [37] survival_2.43-3         iterators_1.0.10       
##  [39] ape_5.2                 glue_1.3.0             
##  [41] registry_0.5-1          gtable_0.2.0           
##  [43] zlibbioc_1.28.0         XVector_0.22.0         
##  [45] webshot_0.5.1           kernlab_0.9-27         
##  [47] Rhdf5lib_1.4.2          HDF5Array_1.10.1       
##  [49] prabclus_2.2-7          DEoptimR_1.0-8         
##  [51] scales_1.0.0            mvtnorm_1.0-10         
##  [53] DBI_1.0.0               edgeR_3.24.3           
##  [55] rngtools_1.3.1          bibtex_0.4.2           
##  [57] miniUI_0.1.1.1          Rcpp_1.0.0             
##  [59] viridisLite_0.3.0       xtable_1.8-3           
##  [61] progress_1.2.0          bit_1.1-14             
##  [63] mclust_5.4.2            glmnet_2.0-16          
##  [65] htmlwidgets_1.3         httr_1.4.0             
##  [67] fpc_2.1-11.1            modeltools_0.2-22      
##  [69] pkgconfig_2.0.2         XML_3.98-1.19          
##  [71] flexmix_2.3-15          nnet_7.3-12            
##  [73] locfit_1.5-9.1          labeling_0.3           
##  [75] manipulateWidget_0.10.0 later_0.8.0            
##  [77] howmany_0.3-1           tidyselect_0.2.5       
##  [79] rlang_0.3.1             softImpute_1.4         
##  [81] reshape2_1.4.3          AnnotationDbi_1.44.0   
##  [83] munsell_0.5.0           tools_3.5.0            
##  [85] RSQLite_2.1.1           ade4_1.7-13            
##  [87] evaluate_0.13           stringr_1.4.0          
##  [89] yaml_2.2.0              knitr_1.22             
##  [91] bit64_0.9-7             fs_1.2.6               
##  [93] robustbase_0.93-3       rgl_0.99.16            
##  [95] purrr_0.3.1             dendextend_1.10.0      
##  [97] mime_0.6                whisker_0.3-2          
##  [99] xml2_1.2.0              compiler_3.5.0         
## [101] rstudioapi_0.9.0        tibble_2.0.1           
## [103] RNeXML_2.3.0            pcaPP_1.9-73           
## [105] stringi_1.4.3           gsl_1.9-10.3           
## [107] desc_1.2.0              RSpectra_0.13-1        
## [109] lattice_0.20-38         trimcluster_0.1-2.1    
## [111] Matrix_1.2-16           commonmark_1.7         
## [113] pillar_1.3.1            ADGofTest_0.3          
## [115] zinbwave_1.4.2          bitops_1.0-6           
## [117] httpuv_1.4.5.1          R6_2.4.0               
## [119] promises_1.0.1          gridExtra_2.3          
## [121] codetools_0.2-16        MASS_7.3-51.1          
## [123] assertthat_0.2.0        rhdf5_2.26.2           
## [125] pkgmaker_0.27           rprojroot_1.3-2        
## [127] withr_2.1.2             GenomeInfoDbData_1.2.0 
## [129] locfdr_1.1-8            diptest_0.75-7         
## [131] hms_0.4.2               grid_3.5.0             
## [133] class_7.3-15            rmarkdown_1.12         
## [135] shiny_1.2.0             numDeriv_2016.8-1      
## [137] lubridate_1.7.4</code></pre>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<div id="refs" class="references">
<div id="ref-Paul2015">
<p>Paul, Franziska, Ya’ara Arkin, Amir Giladi, DiegoÂ Adhemar Jaitin, Ephraim Kenigsberg, Hadas Keren-Shaul, Deborah Winter, et al. 2015. “Transcriptional Heterogeneity and Lineage Commitment in Myeloid Progenitors.” <em>Cell</em> 163 (7). Cell Press: 1663–77. doi:<a href="https://doi.org/10.1016/J.CELL.2015.11.013">10.1016/J.CELL.2015.11.013</a>.</p>
</div>
<div id="ref-Risso2018">
<p>Risso, Davide, Liam Purvis, Russell B. Fletcher, Diya Das, John Ngai, Sandrine Dudoit, and Elizabeth Purdom. 2018. “clusterExperiment and RSEC: A Bioconductor package and framework for clustering of single-cell and other large gene expression datasets.” Edited by Aaron E. Darling. <em>PLOS Computational Biology</em> 14 (9). Public Library of Science: e1006378. doi:<a href="https://doi.org/10.1371/journal.pcbi.1006378">10.1371/journal.pcbi.1006378</a>.</p>
</div>
<div id="ref-Robinson2010">
<p>Robinson, Mark D, and Alicia Oshlack. 2010. “A scaling normalization method for differential expression analysis of RNA-seq data.” <em>Genome Biology</em> 11 (3). BioMed Central: R25. doi:<a href="https://doi.org/10.1186/gb-2010-11-3-r25">10.1186/gb-2010-11-3-r25</a>.</p>
</div>
<div id="ref-Street2018a">
<p>Street, Kelly, Davide Risso, Russell B. Fletcher, Diya Das, John Ngai, Nir Yosef, Elizabeth Purdom, and Sandrine Dudoit. 2018. “Slingshot: cell lineage and pseudotime inference for single-cell transcriptomics.” <em>BMC Genomics</em> 19 (1). BioMed Central: 477. doi:<a href="https://doi.org/10.1186/s12864-018-4772-0">10.1186/s12864-018-4772-0</a>.</p>
</div>
<div id="ref-VandenBerge2018">
<p>Van den Berge, Koen, Fanny Perraudeau, Charlotte Soneson, Michael I. Love, Davide Risso, Jean-Philippe Vert, Mark D. Robinson, Sandrine Dudoit, and Lieven Clement. 2018. “Observation weights unlock bulk RNA-seq tools for zero inflation and single-cell applications.” <em>Genome Biology</em> 19 (1). BioMed Central: 24. doi:<a href="https://doi.org/10.1186/s13059-018-1406-4">10.1186/s13059-018-1406-4</a>.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#load-data">Load data</a></li>
      <li><a href="#fit-trajectories-using-slingshot">Fit trajectories using slingshot</a></li>
      <li><a href="#fit-additive-models">Fit additive models</a></li>
      <li>
<a href="#within-lineage-comparisons">Within-lineage comparisons</a><ul class="nav nav-pills nav-stacked">
<li><a href="#association-of-gene-expression-with-pseudotime">Association of gene expression with pseudotime</a></li>
      <li><a href="#discovering-progenitor-marker-genes">Discovering progenitor marker genes</a></li>
      </ul>
</li>
      <li>
<a href="#between-lineage-comparisons">Between-lineage comparisons</a><ul class="nav nav-pills nav-stacked">
<li><a href="#discovering-differentiated-cell-type-markers">Discovering differentiated cell type markers</a></li>
      <li><a href="#discovering-genes-with-different-expression-patterns">Discovering genes with different expression patterns</a></li>
      <li><a href="#early-drivers-of-differentiation">Early drivers of differentiation</a></li>
      </ul>
</li>
      <li><a href="#clustering-of-genes-according-to-their-expression-pattern">Clustering of genes according to their expression pattern</a></li>
      <li><a href="#cheatsheet">Cheatsheet</a></li>
      <li><a href="#session">Session</a></li>
      <li><a href="#references">References</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Koen Van den Berge, Hector Roux de Bézieux.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
